---
title: "A BASiCS workflow for expression variability analysis using scRNAseq data"
author:
- name: Nils Eling
  affiliation: 
  - &EBI European Molecular Biology Laboratory, European Bioinformatics Institute (EMBL-EBI), Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
  - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Cambridge, CB2 0RE, UK
- name: John C. Marioni
  affiliation: 
  - *EBI
  - *CRUK
- name: Catalina A. Vallejos
  affiliation: 
  - The Alan Turing Institute, British Library, 96 Euston Road, London NW1 2DB, UK
  - Department of Statistical Science, University College London, 1-19 Torrington Place, London WC1E 7HB, UK
  email: INSERT.EMAIL
abstract: Abstracts should be up to 300 words and provide a succinct summary of the article. Although the abstract should explain why the article might be interesting, care should be taken not to inappropriately over-emphasise the importance of the work described in the article. Citations should not be used in the abstract, and the use of abbreviations should be minimized.
keywords: Please list up to eight keywords to help readers interested in your article find it more easily.
bibliography: BASiCSworkflow.bib
output: BiocWorkflowTools::f1000_article
#output:
# BiocStyle::html_document:
#   fig_caption: true
#   self_contained: yes
#vignette: >
#  %\VignetteIndexEntry{A BASiCS workflow for expression variability analysis using scRNAseq data}
#  %\VignetteEngine{knitr::rmarkdown}
#  %\VignetteEncoding{UTF-8}
---

```{r setup_knitr, include = FALSE, cache = FALSE}
#library(BiocStyle)
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
#on.bioc <- TRUE
#library(knitr)
# Use fig.width = 7 for html and fig.width = 6 for pdf
#fig.width <- ifelse(on.bioc, 7, 6)
#knitr::opts_chunk$set(cache = 2, warning = FALSE, message = FALSE, error = FALSE,
#  cache.path = "cache/", fig.path = "figure/", fig.width = fig.width)
```

# Introduction

Single-cell RNA-sequencing (scRNA-seq) is the leading technology to study transcriptional heterogeneity in cellular subpopulations that remains undetected in bulk experiments [@Stegle2015]. Its applications range from studying cell types in immunity [@Lonnberg2017] and development [@Ibarra-Soria2018a] to dissecting the mechanism for cell-to-cell variability in expression [@Goolam2016] or understanding differentiation processes [@Proserpio2016]. Within a population of cells, transcript detection differences can be regarded as a mixture of possible causes. Extrinsic biological noise in form of differences in dynamic states (cell size, cell cycle, metabolism, intra- and inter-cellular signalling) and intrinsic noise that effects single promoters (mutations and epigenetic tags) alter gene expression within the cell [@Eberwine2015] while mRNA capture inefficiencies distort the detected transcript counts (technical noise) [@Brennecke2013].

To account for high amounts of technical noise that affects scRNA-seq data, external RNA spike-in molecules (ERCCs) [@Rna2005] can be added to the experiment before sequencing. Fitting the variance-mean dependency of the ERCCs allows the statistical detection of genes with higher dispersion in expression than the technical background [@Brennecke2013]. This idea has been expanded to estimate cell specific normalisation factors in parallel within the BASiCS model. Total variability can be decomposed into a technical and biological part by incorporating ERCC reads [@Vallejos2015]. Since the era of bulk RNA sequencing, methods for differential expression testing of transcript counts across two conditions have been developed [@Anders2010;@Robinson2009]. Due to zero inflation in scRNA-seq data, approaches different from bulk RNA-seq analysis were developed for differential expression testing in scRNA-seq data [@Katayama2013;@Kharchenko2014;@Delmans2016]. In contrast to bulk samples, scRNA-seq allows the characterization of gene expression variability across a population of cells. To compare changes in transcriptional variability, the BASiCS package has been expanded to include differential mean expression and differential variability testing. In this setting, BASiCS uses the over-dispersion after correcting for technical noise as proxy for biolgocial variability [@Vallejos2016a]. Due to a strong confounding between mean expression and over-dispersion, BASiCS has been extended to model this relationship an capture residual over-dispersion estimates that show no correlation to mean expression. Therefore, it is possible to test changes in mean expression in parallel to changes in variability [@Eling2017]. 

Here we present a computational workflow for differential expression and differential variability testing of scRNA-seq data using the previously published BASiCS package [@Vallejos2015;@Vallejos2016a;@Eling2017] implemented in the statistical computing language R. The analysis starts with a raw transcript counts matrix after removal of low-quality cells. A full quality control process is explained elsewhere [@McCarthy2017]. We will first focus on gene filtering, generating a BASiCS Data object, running and testing Markov-Chain Monte Carlo simulations. Additionally, we will use two examples (Cel-Seq [@Grun2014] and droplet-based scRNAseq data [@Ibarra-Soria2018a]) to exemplify statistical detection of highly variable, differentially expressed and differentially variable genes in different conditions. 

# Methods

## The BASiCS model
The BASiCS package features a hierarchical Bayesian model in which the transcript count of gene $i$ of cell $j$ is represented as a random variable $X_{ij}$. For all technical genes, counts can be modelled as Poisson distributed with a random effect capturing unexplained technical noise:

$$
\begin{aligned}
X_{ij}&|\mu_i,\nu_j\sim{}Poisson(\nu_j\mu_i)\\
\nu_j&|s_j,\theta\sim{}Gamma(1/\theta,1/(s_j\theta))
\end{aligned}
$$

where $\mu_i$ explains the gene's mean expression, $\nu_j$ the technical effect dominated by the mRNA capture efficiency $s_j$ and the unexplained technical noise parameter $\theta$. 

This technical influence effects all biological genes in the same manner. In addition to technical variations, cell-to-cell differences of biological gene expression can be cause by biological processes as explained above. To model an additional random effect that captures biological cell-to-cell variability, the random variable $X_{ij}$ is modelled as: 

$$
\begin{aligned}
X_{ij}&|\mu_i,\phi_j,\nu_j,\rho_{ij}\sim{}Poisson(\phi_j\nu_j\mu_i\rho_{ij})\\
\rho_{ij}&|\delta_i\sim{}Gamma(1/\delta_i,1/\delta_i)
\end{aligned}
$$

where $\phi_j$ is the cell-specifc size factor and $\rho_{ij}$ the biological random effect incoroporating the over-dispersion hyper-parameter $\delta_i$.

In the further analysis, $\mu_i$ represents the mean expression and $\delta_i$ the biological over-dispersion of each gene.

### Testing for changes in mean expression and over-dispersion

Due to a strong confounding that is typically observed between mean and over-dispersion for scRNA-seq datasets, 

### Correcting the mean-variability confounding effect


## Operation
This part of the methods should include the minimal system requirements needed to run the software and an overview of the workflow for the tool for users of the tool.


# Analysis of serum grown mESCs (single group example)

Give a short description of the data

## Obtaining the data

Use direct download link

## Filtering the data

## Caluclating the molecule count for ERCC spike ins

Use one of the condition in Grun

## Creating the BASiCS Data object

## Running the MCMC

## Downstream analysis

### Lowly-highly variable genes

# Differential testing using mESCs in two conditions (two group example)

## Obtaining the data

Shorter description as above

## Filtering the data

## Creating the BASiCS Data objects

## Running the MCMC

## Downstream analysis

### Differential mean expression

### Differential over-dispersion

### Differential residual over-dispersion

# Differential testing using differentiating cells (two group, droplet-based example)


# Summary 
This section is required if the paper does not include novel data or analyses.  It allows authors to briefly summarize the key points from the article.


# Software availability
This section will be generated by the Editorial Office before publication. Authors are asked to provide some initial information to assist the Editorial Office, as detailed below.

1. URL link to where the software can be downloaded from or used by a non-coder (AUTHOR TO PROVIDE; optional)
2. URL link to the author's version control system repository containing the source code (AUTHOR TO PROVIDE; required)
3. Link to source code as at time of publication (*F1000Research* TO GENERATE)
4. Link to archived source code as at time of publication (*F1000Research* TO GENERATE)
5. Software license (AUTHOR TO PROVIDE; required)


# Author information
In order to give appropriate credit to each author of an article, the individual contributions of each author to the manuscript should be detailed in this section. We recommend using author initials and then stating briefly how they contributed.

# Competing interests
All financial, personal, or professional competing interests for any of the authors that could be construed to unduly influence the content of the article must be disclosed and will be displayed alongside the article. If there are no relevant competing interests to declare, please add the following: 'No competing interests were disclosed'.

# Grant information
Please state who funded the work discussed in this article, whether it is your employer, a grant funder etc. Please do not list funding that you have that is not relevant to this specific piece of research. For each funder, please state the funderâ€™s name, the grant number where applicable, and the individual to whom the grant was assigned. If your work was not funded by any grants, please include the line: 'The author(s) declared that no grants were involved in supporting this work.'

# Acknowledgments
This section should acknowledge anyone who contributed to the research or the article but who does not qualify as an author based on the criteria provided earlier (e.g. someone or an organization that provided writing assistance). Please state how they contributed; authors should obtain permission to acknowledge from all those mentioned in the Acknowledgments section.

Please do not list grant funding in this section.