---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Nils Eling
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
    email: eling@ebi.ac.uk
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
  - name: John C. Marioni
    affiliation: 
    - *EBI
    - *CRUK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems. Single-cell RNA sequencing can be used to quantify this 
  heterogeneity, but it is prone to strong technical noise. Here, we describe a 
  step-by-step computational workflow which uses the BASiCS Bioconductor package 
  to robustly quantify expression variability within and between known cell 
  populations (such as experimental conditions or cell types). BASiCS provides 
  an integrated framework for data normalisation, technical noise quantification
  and downstream analyses, whilst propagating statistical uncertainty across 
  these steps. Within a single seemingly homogeneous cell population, BASiCS 
  can be used identify highly variable genes that drive the heterogeneity 
  within the population as well as lowly variable genes that might exhibit 
  housekeeping-like behavior. BASiCS also provides a probabilistic rule to
  identify changes in expression variability between cell populations, while 
  avoiding confounding effects related to differences in technical noise or in 
  overall abundance. Using two publicly available datasets, we guide users 
  through a complete pipeline which includes preliminary steps for quality 
  control and data exploration using the scater and scran Bioconductor packages. 
  Data for the first case study was generated using the Fluidigm\@ C1 system, in 
  which extrinsic spike-in RNA molecules were added in order to quantify 
  technical noise. The second dataset was generated using a droplet-based 
  system, for which spike-in RNA is not available. The latter analysis provides 
  an example, in which differential variability testing reveals insights 
  regarding a possible early cell fate commitment process.
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article: default
  BiocStyle::html_document: default
---


```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
# Use fig.width = 7 for html and fig.width = 6 for pdf
fig.width <- ifelse(on.bioc, 7, 6)
knitr::opts_chunk$set(
  cache = TRUE, warning = FALSE, message = FALSE, error = FALSE,
  cache.path = "cache/", fig.path = "figure/", fig.width = fig.width
)
```


# Introduction

<!--- scRNA-Seq and the different types of heterogeneity ---> 
Single-cell RNA-sequencing (scRNA-seq) enables the study of genome-wide 
transcriptional heterogeneity in cell populations that remains otherwise 
undetected in bulk experiments [@Stegle2015; @Prakadan2017; @Patange2018]. 
Existing applications range from characterising cell types in immunity 
[@Lonnberg2017;  @Villani2017; @Zheng2017] and development [@Ibarra-Soria2018; 
@Wagner2018; @Pijuan-Sala2019] to dissecting the mechanisms for cell fate 
commitment [@Goolam2016; @Ohnishi2014]. 
Transcriptional heterogeneity within a population of cells can relate to 
different reasons.
On the broadest level, this heterogeneity can reflect the presence of distinct 
expression profiles associated to cell subtypes or states which could be 
characterised through clustering [@Kiselev2019]. 
<!-- something about pseudotime trajectories --> 
More subtle gene expression variability within a seemingly homogeneous cell 
population can be due to deterministic or stochastic events, the later being 
referred to as 'noise' [@Elowitz2002]. 