---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Nils Eling
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
    email: eling@ebi.ac.uk
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
  - name: John C. Marioni
    affiliation: 
    - *EBI
    - *CRUK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems. Single-cell RNA sequencing can be used to quantify this 
  heterogeneity, but it is prone to strong technical noise. Here, we describe a 
  step-by-step computational workflow which uses the BASiCS Bioconductor package 
  to robustly quantify expression variability within and between known cell 
  populations (such as experimental conditions or cell types). BASiCS provides 
  an integrated framework for data normalisation, technical noise quantification
  and downstream analyses, whilst propagating statistical uncertainty across 
  these steps. Within a single seemingly homogeneous cell population, BASiCS 
  can be used identify highly variable genes that drive the heterogeneity 
  within the population as well as lowly variable genes that might exhibit 
  housekeeping-like behavior. BASiCS also provides a probabilistic rule to
  identify changes in expression variability between cell populations, while 
  avoiding confounding effects related to differences in technical noise or in 
  overall abundance. Using two publicly available datasets, we guide users 
  through a complete pipeline which includes preliminary steps for quality 
  control and data exploration using the scater and scran Bioconductor packages. 
  Data for the first case study was generated using the Fluidigm\@ C1 system, in 
  which extrinsic spike-in RNA molecules were added in order to quantify 
  technical noise. The second dataset was generated using a droplet-based 
  system, for which spike-in RNA is not available. The latter analysis provides 
  an example, in which differential variability testing reveals insights 
  regarding a possible early cell fate commitment process.
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article: default
  BiocStyle::html_document: default
---


```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
# Use fig.width = 7 for html and fig.width = 6 for pdf
fig.width <- ifelse(on.bioc, 7, 6)
knitr::opts_chunk$set(
  cache = TRUE, warning = FALSE, message = FALSE, error = FALSE,
  cache.path = "cache/", fig.path = "figure/", fig.width = fig.width
)
```


# Introduction

<!--- scRNA-Seq and the different types of heterogeneity ---> 
Single-cell RNA-sequencing (scRNA-seq) enables the study of genome-wide 
transcriptional heterogeneity in cell populations that remains otherwise 
undetected in bulk experiments [@Stegle2015; @Prakadan2017; @Patange2018]. 
Applications of scRNA-seq range from characterising cell types in immunity 
[@Lonnberg2017;  @Villani2017; @Zheng2017] and development [@Ibarra-Soria2018; 
@Wagner2018; @Pijuan-Sala2019] to dissecting the mechanisms for cell fate 
commitment [@Goolam2016; @Ohnishi2014]. 
Transcriptional heterogeneity within a cell population can relate to different 
underlying sources.
On the broadest level, this heterogeneity can reflect the presence of distinct 
expression profiles associated to cell subtypes or discrete states which could 
be identified through clustering [@Kiselev2019]. 
Alternatively, cell-to-cell expression heterogeneity can be due to gradual 
changes along biological processes that evolve over time (such as 
development and differentiation) --- these can be characterised using pseudotime
inference methods [@Saelens2019].
More subtle expression variability within a seemingly homogeneous cell 
population can be due to deterministic or stochastic events and is the focus 
of this article. 
The stochastic component of this variability is referred to as transcriptional 
*noise* [@Elowitz2002; @Eling2019]. 

<!--- why is it challenging to study transcriptional noise? ---> 
Transcriptional noise can arise from intrinsic and extrinsic sources of 
variability. Classically, extrinsic noise is defined as stochastic fluctuations 
in cellular components, which is induced by cells residing in different dynamic 
states (e.g. cell size, cell cycle, metabolism, intra- and inter-cellular 
signalling) [@Zopf2013; @Iwamoto2016; @Kiviet2014]. 
Instead, intrinsic noise arises from stochastic effects on biochemical 
processes such as transcription and translation [@Elowitz2002]. 
Intrinsic noise can be modulated by genetic and epigenetic modifications (such 
as mutations, histone modifications, CpG island length and nucleosome 
positioning) [@Eberwine2015; @Faure2017; @Morgan2018] and is usually measured 
at the level of individual genes [@Elowitz2002]. Cell-to-cell gene expression 
variability estimates derived from scRNA-seq data capture a combination of 
these effects, as well as deterministic regulatory mechanisms [@Eling2019]. 
These variability estimates can also be inflated by the technical noise that is 
typically observed in scRNA-seq assays [@Brennecke2013].

Different strategies have been implemented to quantify or attenuate technical
noise in scRNA-seq experiments. For example, external RNA spike-in molecules 
(such as the ones introduced by the External RNA Controls Consortium, ERCCs 
[@Rna2005]) can be added to each cellâ€™s lysate. Spike-ins can be used to inform
quality control steps [@McCarthy2017], data normalisation [@Vallejos2017] as 
well as to infer technical background noise [@Brennecke2013]. 

<!-- next talk about UMIs and computational approaches ---> 

Some computational methods aim to denoise the data prior to downstream analysis
(e.g. via imputation). Alternatively, computational approaches can be designed
to simultaneously quantify technical variabli

was to quantify or attenuate technical noise in scRNA-seq assays. 
Some are experimental: eg spike-ins or UMIs. Some others are computational. 

<!--- Technical noise, normalisation and hvg/lvg analyses ---> 
Moreover, technical noise inflates the observed cell-to-cell variability in gene
expression [@Brennecke2013].
To account for high amounts of technical noise that affects scRNA-seq data, 

Fitting a regression trend between the variability and the mean abundance of the 
ERCC molecules allows the statistical detection of genes the show larger 
variability than the technical background . 

Genes that show larger variability compared to spike-in molecules or the average 
variability are often referred to as 'highly variable genes' (HVG) and are used in 
computational scRNA-Seq analysis to select biologically informative genes for 
down-stream analysis [@Lun2016].
Furthermore, spike-in molecules can be used to normalize gene expression for 
cells with differences in total mRNA content.