---
title: "Supplementary material for: A BASiCS workflow for expression variability analysis using scRNA-Seq data"
author: 
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
    email: "a.b.o'callaghan@sms.ed.ac.uk"
  - name: Nils Eling
    affiliation: 
    - &UZH Department of Quantitative Biomedicine, University of Zurich,
      Winterthurerstrasse 190, CH-8057, Zurich, Switzerland
    - &ETH Institute for Molecular Health Sciences, ETH Zurich,
      Otto-Stern Weg 7, 8093 Zurich, Switzerland
  - name: John C. Marioni
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article:
    fig_width: 6
    fig_height: 4
---


```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
library("knitr")
knitr::opts_chunk$set(
  message = FALSE, error = FALSE, warning = FALSE,
  # cache = 2, cache.path = "cache_supp/",
  fig.path = "figure_supp/"
)
```

# Methods

## The BASiCS model

The BASiCS package features a hierarchical Bayesian model in which the 
transcript count of gene $i$ of cell $j$ is represented as a random variable
$X_{ij}$. For all technical genes, counts can be modelled as Poisson distributed
with a random effect capturing unexplained technical noise:
$$
\begin{aligned}
  X_{ij}|\mu_i,\nu_j & \sim \text{Poisson}(\nu_j\mu_i)\\
  \nu_j|s_j,\theta & \sim \text{Gamma}\left(\frac{1}{\theta},\frac{1}{s_j\theta}\right)
\end{aligned}
$$
where $\mu_i$ explains the gene's mean expression, $\nu_j$ the technical effect
dominated by the mRNA capture efficiency $s_j$ and the unexplained technical 
noise parameter $\theta$. 

This technical influence effects all biological genes in the same manner. 
In addition to technical variations, cell-to-cell differences of biological 
gene expression can be cause by biological processes as explained above. 
To model an additional random effect that captures biological cell-to-cell 
variability, the random variable $X_{ij}$ is modelled as: 
$$
\begin{aligned}
  X_{ij}|\mu_i,\phi_j,\nu_j,\rho_{ij} & \sim \text{Poisson}(\phi_j\nu_j\mu_i\rho_{ij})\\
  \rho_{ij}|\delta_i & \sim \text{Gamma}(\frac{1}{\delta_i}, \frac{1}{\delta_i})
\end{aligned}
$$
where $\phi_j$ is the cell-specifc size factor and $\rho_{ij}$ the biological
random effect incorporating the over-dispersion hyper-parameter $\delta_i$.

In further analysis, $\mu_i$ represents the mean expression and $\delta_i$
the biological over-dispersion of each gene.

### Testing for changes in mean expression and over-dispersion

Differential mean and differential over-dispersion testing is done by computing
the tail posterior propabilities of the difference in mean expression or 
over-dispersion between two conditions ($A$ and $B$) being larger than an
evidence threshold $\tau_0$ or $\omega_0$ [@Bochkina2007, @Vallejos2016]:
$$
\begin{aligned}
  &\text{P}\left(\log\left(\frac{\mu_i^{(A)}}{\mu_i^{(B)}}\right) > \tau_0 | \text{Data}\right) > \alpha_m\\
  &\text{P}\left(\log\left(\frac{\delta_i^{(A)}}{\delta_i^{(B)}}\right) > \omega_0 | \text{Data}\right) > \alpha_d
\end{aligned}
$$

If the tail posterior probability is larger than a given propability threshold 
$\alpha_m$ or $\alpha_d$, the gene is considered to be differentially expressed
or differentially over-dispersed [@Vallejos2016]. The evidence threshold is
usually fixed *a priori* and the probability threshold is defined to
control the expected false discovery rate (EFDR) to 
(e.g. 10%) [@Newton2004, Vallejos2016].

As described by @Vallejos2016, estimates of the
over-dispersion parameters $\delta_i$ are negatively correlated to mean 
expression $\mu_i$. This indicates that in homogeneous populations of cells,
highly expressed genes tend to be less noisy than lowly expressed genes. 
Differential over-dispersion testing is therefore confounded by mean expression
changes. When assessing changes in over-dispersion $\delta_i$, only genes with
no changes in mean expression are considered (see @Vallejos2016).

### Correcting the mean-variability confounding effect

@Eling2018 extended BASiCS to account for the
confounding effect between mean expression and expression variability. 
For this purpose, we capture the relationship between mean and over-dispersion
parameters by introducing the following joint prior distribution for 
$(\mu_i, \delta_i)'$:
$$
  \mu_i \sim \text{log-Normal}\left(0, s^2_{\mu}\right), \hspace{0.5cm}
  \delta_i | \mu_i \sim \text{log-t}_{\eta}\left( \text{f}(\mu_i), \sigma^2 \right).
$$

The latter is equivalent to the non-linear regression model:
$$
  \log(\delta_i) = \text{f}(\mu_i) + \epsilon_i, \hspace{0.5cm} \epsilon_i \sim{}\text{t}_{\eta}(0,\sigma^2), 
$$

where $\mbox{f}(\mu_i)$ represents the over-dispersion (on the log-scale) that
is predicted by the global trend (across all genes) for a given mean expression 
$\mu_i$. Therefore, $\epsilon_i$ can be interpreted as a gene-specific 
*residual over-dispersion* parameter. Positive values for
$\epsilon_i$ indicate more variation than expected for genes with similar 
expression level. Similarly, negative values of $\epsilon_i$ suggest less 
variation than expected [@Eling2018].

In line with the probabilistic approach described above, we identified
statistically significant differences in residual over-dispersion for those 
genes where the tail posterior probability of observing a large difference 
between $\epsilon_i^A$ and $\epsilon_i^B$ exceeds a certain threshold:
$$
  \text{P}(\mid\epsilon_i^{A} - \epsilon_i^{B}\mid > \psi_0 \mid \text{Data}) > \alpha_R,
$$
where $\psi_0 > 0$ defines the minimum tolerance threshold. 
As a default choice, we assume $\psi_0 = \log_2(1.5) / \log_2(e) \approx 0.41$,
which translates into a 50\% increase in over-dispersion. 
The posterior probability threshold $\alpha_R$ is chosen to control the EFDR
(e.g. 10\%) [@Newton2004]. To support interpretability of the results,
we exclude genes that are not expressed in at least 2 cells per condition from
differential variability testing.

# Trace plots

First, we will load `r Biocpkg("BASiCS")` and read in each MCMC chain.
While it is important to ensure the MCMC algorithm has converged and has
produced an adequate number of independent samples from the posterior
distribution, we will here assume this proceeds as in the MCMC diagnostics
section of the main manuscript.

```{r load-chain-naive}
library("BASiCS")
MCMC_naive <- readRDS("rds/chain_naive.Rds")
MCMC_active <- readRDS("rds/chain_active.Rds")
MCMC_naive_nospikes <- readRDS("rds/chain_naive_nospikes.Rds")
```

## Global scaling normalisation using BASiCS

Global scaling normalisation techniques for scRNA-seq data can be used to infer
cell-specific scaling factors that capture systematic differences in the scale
of the expression counts across cells. 
Generally, these scaling factors capture both biological (e.g. mRNA content) and technical (e.g. sequencing depth) factors, but spike-in genes can be used to
disentangle these effects [@Vallejos2017]. 
For example, when spike-in genes are used to assist data normalisation, 
`r Biocpkg("BASiCS")` infers two separate sets of cell-specific parameters 
associated to these sources of variability: $\phi_j$ and $\nu_j$, respectively.  
Posterior estimates for these parameters ($\hat{\phi}_j$, $\hat{\nu}_j$) can be 
used to normalise the input data via the `BASiCS_DenoisedCounts` function. 
This function returns normalised expression values defined as
$$
  x^*_{ij} = \frac{ x_{ij} } {\hat{\phi}_j \hat{\nu}_j},
$$
where $x_{ij}$ denotes the observed expression count for gene $i$ in cell $j$.

```{r normalisation-naive}
## to calculate normalised counts we need the original SCE objects
## used to create the MCMC objects
sce_naive <- readRDS("rds/sce_naive.rds")
sce_active <- readRDS("rds/sce_active.rds")

## First remove global differences in scale for normalisation factors
offset_corrected <- BASiCS_CorrectOffset(MCMC_active, MCMC_naive)
norm_basics_naive <- BASiCS_DenoisedCounts(
  Data = sce_naive,
  Chain = MCMC_naive
)
norm_basics_active <- BASiCS_DenoisedCounts(
  Data = sce_active,
  Chain = offset_corrected$Chain
)
```

As shown in Figure \@ref(fig:normalisation-vs-scran), the capture efficiency and 
scaling factors ($\nu_j$ and $\phi_j$ respectively) inferred by 
`r Biocpkg("BASiCS")` are highly correlated with the global scaling 
normalisation factors inferred using the method of @Lun2016.
We will load `r Biocpkg("scran")` to estimate these size factors.
An advantage of jointly inferring normalisation and gene-specific expression
parameters, rather than using pre-computed
normalisation factors, is that in the former case, statistical uncertainty is 
propagated from normalisation to all downstream analyses.

```{r normalisation-vs-scran, fig.width=4, fig.height=5, fig.cap="Comparison of BASiCS inferred global scaling factors and those estimated by scran using the method of Lun et al. (2016)."}
## load scran to estimate size factors and ggplot2 to make graphs
library("scran")
library("ggplot2")
## set default theme for plots
theme_set(theme_bw())

## BASiCS scaling factors within each group of samples
basics_sf_naive <- colMedians(
  displayChainBASiCS(MCMC_naive, "phi")
  * displayChainBASiCS(MCMC_naive, "nu")
)
basics_sf_active <- colMedians(
  displayChainBASiCS(MCMC_active, "phi")
  * displayChainBASiCS(MCMC_active, "nu")
)
## scran scaling factors within each group of samples
scran_sf_naive <- calculateSumFactors(sce_naive)
scran_sf_active <- calculateSumFactors(sce_active)

ggplot() +
  aes(basics_sf_naive, scran_sf_naive) +
  geom_point() +
  labs(x = "BASiCS scaling factors", y = "scran scaling factors") +
  annotate(
    x = 14,
    y = 0.4,
    label = paste(
      "Pearson's r = ",
      round(
        cor(basics_sf_naive, scran_sf_naive),
        digits = 2
      )
    ),
    geom = "text",
    size = 4
  )
```


# Differential testing using droplet-based differentiating cells {#droplet-data}

The development of droplet-based scRNA-seq [@Klein2015;@Macosko2015] 
lead to a large increase in the number of cells that can be profiled per 
experiment at a fixed cost.
With this, large-scale scRNA-seq datasets have been generated to study 
development across multiple time-points and capturing multiple tissues 
[@Ibarra-Soria2018;@Kernfeld2018]. 
Here we describe the computational analysis of changes in mean expression 
and transcriptional variability when data are sparse and technical spike-in 
genes are missing. 
For this, we compare cells of the presomitic mesoderm and somitic mesoderm 
using droplet-based scRNA-seq data [@Ibarra-Soria2018].

## Obtaining the data

The full dataset is stored under the accession number E-MTAB-6153 on 
ArrayExpress and can be obtained via:

```{r load-droplet-data}
options(timeout = 1000)
if (!file.exists("downloads/rawCounts.tsv")) {
  website <- "https://www.ebi.ac.uk/"
  folder <- "arrayexpress/files/E-MTAB-6153/"
  file <- "E-MTAB-6153.processed.2.zip"
  download.file(
    paste0(website, folder, file),
    destfile = "downloads/rawCounts.zip"
  )
  unzip(zipfile = "downloads/rawCounts.zip", exdir = "downloads")
  file.remove("downloads/rawCounts.zip")
}
raw_counts <- read.delim("downloads/rawCounts.tsv", header = TRUE)
```

*Of note:* The zip file is 65 MB in size while the unzipped, raw counts measure
873 MB in size. The cluster labels of the original publication ca be obtained 
via:

```{r cluster-labels-droplet}
if (!file.exists("downloads/cellAnnotation.tsv")) {
  website <- "https://www.ebi.ac.uk/"
  folder <- "arrayexpress/files/E-MTAB-6153/"
  file <- "E-MTAB-6153.processed.3.zip"
  download.file(
    paste0(website, folder, file),
    destfile = "cluster_labels.zip"
  )
  unzip(zipfile = "cluster_labels.zip", exdir = "downloads")
}

cluster_labels <- read.table("downloads/cellAnnotation.tsv",
  sep = "\t", header = TRUE, stringsAsFactors = FALSE)
```

## Select populations of interest

We select the somitic and pre-somitic mesoderm cells to perform differential 
testing. Prior to running the MCMC, we want to control for outlying cells and 
heterogeneous substructure in both cell populations. 

```{r droplet-cell-selection}
cluster_labels[["Cell_type"]] <- cluster_labels$cellType
cluster_labels[["Cell_type"]] <- sub(
  "presomiticMesoderm",
  "PSM",
  cluster_labels[["Cell_type"]]
)
cluster_labels[["Cell_type"]] <- sub("somiticMesoderm",
  "SM",
  cluster_labels[["Cell_type"]]
)

ind_som <- which(cluster_labels[["Cell_type"]] == "PSM" |
  cluster_labels[["Cell_type"]] == "SM")
raw_counts <- raw_counts[, ind_som]
cluster_labels <- cluster_labels[ind_som, ]
```

## Generating SingleCellExperiment object and quality control

For pre-processing and visualisation purposes, we load the data into a 
SingleCellExperiment object. The metadata will be stored in the `colData` slot.

```{r droplet-sce}
droplet_sce <- SingleCellExperiment(
  assays = list(counts = as(as.matrix(raw_counts), "dgCMatrix")),
  colData = DataFrame(
    cluster_labels,
    subCellType = sub("_.*", "", cluster_labels$cell)
  )
)
rm(raw_counts)
```

For further processing steps, we remove lowly expressed genes.

```{r filter-drop}
ind_expressed <- Matrix::rowMeans(counts(droplet_sce)) > 0.1
droplet_sce <- droplet_sce[ind_expressed, ]
```

To visualise possible sub-structure in the data, we normalise both cell 
populations using the `scran` package.

```{r norm-drop}
library("scran")
library("scater")
droplet_sce <- computeSumFactors(droplet_sce,
  clusters = colData(droplet_sce)$subCellType
)
droplet_sce <- logNormCounts(droplet_sce)
```

Next, we compute a PCA using the `scater` package. 

```{r droplet-pca}
droplet_sce <- runPCA(droplet_sce)
```

We can now visualise the different factors stored in the `colData` slot.

```{r drop-pca-plot, fig.cap="Principal component plot of pre-somitic and somitic mesoderm cells. Colour indicates cell type."}
## Cell types identified by clustering
plotReducedDim(droplet_sce, dimred = "PCA", colour_by = "subCellType") +
  scale_fill_manual(
    name = "cellType",
    values = c("coral4", "steelblue", "limegreen")
  )
```

The first PC separates the two different cell types while the second PC captures
outlying cells.
We will remove these outliers and the intermediate cell population from 
down-stream analysis.

```{r drop-filt}
ind_retain <- reducedDims(droplet_sce)$PCA[, 2] > -5 &
  colData(droplet_sce)$subCellType != "presomiticMesoderm.b"
droplet_sce <- droplet_sce[, ind_retain]
```

We now collected the cells that we want to process using BASiCS. 
For this, we will generate the BASiCS data objects.

## Generating BASiCS data objects

Since droplet-based scRNA-seq data are generated without including technical 
spike-in genes, BASiCS uses measurement error models to quantify technical 
variation through replication [@Carroll1998].
Here, it is crucial to provide batch information to the BASiCS model.
In the case of the somitic and pre-somoitic mesoderm cells, embryos of two mice 
have been used to generate the data. 
Cells isolated from the first embryo were split into two batches and processed 
independently.
To capture cell-type extrinsic, biological variation between the two mice, we 
pool cells from the two batches of the first animal and only considere cells 
from mouse 1 and mouse 2 as replicates.

```{r BASiCS-droplet-data}
## Presomitic mesoderm
ind_presom <- colData(droplet_sce)[["Cell_type"]] == "PSM"
droplet_sce$batch <- round(colData(droplet_sce)$sample)
PSM_Data <- droplet_sce[, ind_presom]

PSM_Data$BatchInfo <- factor(PSM_Data$batch)

## Somitic mesoderm
ind_som <- colData(droplet_sce)[["Cell_type"]] == "SM"
SM_Data <- droplet_sce[, ind_som]
SM_Data$BatchInfo <- factor(SM_Data$batch)
```

## Running the MCMC

We next estimate model parameters by running the MCMC cell-type specifically. 
Due to the high cell number (1150 for the pre-somitic mesoderm and 739 for the 
somitic mesoderm), we set the number of iterations to 20000. 
In this case, we used the regression BASiCS model to additionally estimate residual 
over-dispersion parameters.

```{r run-mcmc, eval = FALSE}
## Presomitic mesoderm cells
PSM_MCMC <- BASiCS_MCMC(
  PSM_Data,
  N = 20000,
  Thin = 10,
  Burn = 10000,
  Regression = TRUE
)
## Somitic mesoderm cells
SM_MCMC <- BASiCS_MCMC(
  SM_Data,
  N = 20000,
  Thin = 10,
  Burn = 10000,
  Regression = TRUE
)
```

Running these MCMC will take around 8-12 hours on a standard PC (2.6 GHz i5,
8 GB RAM, using 1 core). Here, we provide these chains to download from:

```{r droplet-chain-download}
website <- "https://git.ecdf.ed.ac.uk/"
folder <- "vallejosgroup/basicsworkflow2020/raw/master/"
PSM_MCMC <- readRDS(
  file = url(
    paste0(website, folder, "PSM_MCMC.rds")
  )
)
SM_MCMC <- readRDS(
  file = url(
    paste0(website, folder, "SM_MCMC.rds")
  )
)
```

## Validating the model fit

Next, we visualise the results of the MCMC sampler by visualizing the 
different chains and by plotting the regression trend.
To assess whether the chains converged, we will visualise trace plots for some 
of the parameters. 
The `plot` function allows us to generate trace, posterior density,
and autocorrelation plots for different parameters.

```{r psm-chain-convergence, fig.cap = "Trace plot, marginal density, and autocorrelation function for mean expression of an example gene; presomatic cells."}
plot(PSM_MCMC, Param = "mu", Gene = 120)
```

```{r sm-chain-convergence, fig.cap = "Trace plot, marginal density, and autocorrelation function for mean expression of an example gene; somatic cells."}
plot(SM_MCMC, Param = "mu", Gene = 120)
```


```{r diag-psm, fig.cap="MCMC diagnostics for gene-specific mean expression parameters; presomatic cells. A: Geweke Z-score for mean expression parameters is plotted against mean expression estimates. Dashed lines represent absolute Z-scores of 3, outside of which we advise caution when interpreting results. B: Effective sample size (ESS) is plotted against mean expression estimates. A dashed line shows a threshold of 100, below which we advise caution when interpreting results."}
library("patchwork")
diag_p1 <- BASiCS_DiagPlot(PSM_MCMC, Measure = "geweke")
diag_p1 <- diag_p1 +
  geom_hline(yintercept = c(-3, 3), col = "firebrick", linetype = "dashed") +
  theme(legend.position = "none")

diag_p2 <- BASiCS_DiagPlot(PSM_MCMC, Measure = "ess")
diag_p2 <- diag_p2 +
  geom_hline(yintercept = 100, col = "firebrick", linetype = "dashed") +
  theme(legend.position = "none")

diag_p1 + diag_p2 + plot_annotation(tag_levels = "A")
```


```{r diag-sm, fig.cap="MCMC diagnostics for gene-specific mean expression parameters; somatic cells. A: Geweke Z-score for mean expression parameters is plotted against mean expression estimates. Dashed lines represent absolute Z-scores of 3, outside of which we advise caution when interpreting results. B: Effective sample size (ESS) is plotted against mean expression estimates. A dashed line shows a threshold of 100, below which we advise caution when interpreting results."}
diag_p1 <- BASiCS_DiagPlot(SM_MCMC, Measure = "geweke")
diag_p1 <- diag_p1 +
  geom_hline(yintercept = c(-3, 3), col = "firebrick", linetype = "dashed") +
  theme(legend.position = "none")

diag_p2 <- BASiCS_DiagPlot(SM_MCMC, Measure = "ess")
diag_p2 <- diag_p2 +
  geom_hline(yintercept = 100, col = "firebrick", linetype = "dashed") +
  theme(legend.position = "none")

diag_p1 + diag_p2 + plot_annotation(tag_levels = "A")
```

We observe that the chains for all chosen parameters appear to have converged. 
Furthermore, to validate that the model fitted the mean-variability trend 
correctly, we plot posterior estimates for over-dispersion paramters $\delta_i$ 
against posterior estimates of mean expression parameters $\mu_i$.
For this, the `BASiCS_ShowFit` function can be used.

```{r droplet-regression-trend-psm, fig.cap="Over-dispersion is plotted against mean expression for all genes in pre-somitic mesoderm cells (A) and somitic mesoderm cells (B), on a log-log scale. Colour corresponds to local density, with lighter colours representing higher density. The curve represents the inferred trend of over-dispersion against mean, and the shaded area indicates the 95\\% credible interval."}
a <- BASiCS_ShowFit(PSM_MCMC)
b <- BASiCS_ShowFit(SM_MCMC)

a + b + plot_annotation(tag_levels = "A")
```

Both trends display similar behaviour which allows us to compare residual 
over-dispersion estimates.

## Differential testing 

Next, we test for changes in mean expression and expression variability 
between the somitic and pre-somitic mesoderm.
First, we are interested in assessing global changes in expression 
variability between the two conditions.
For this, we compare over-dispersion parameters $\delta_i$ for genes that are
similarly expressed in both conditions.

```{r drop_testde_lfc0}
Droplet_Test_logFC0 <- BASiCS_TestDE(
  Chain1 = PSM_MCMC,
  Chain2 = SM_MCMC,
  EpsilonM = 0,
  GroupLabel1 = "PSM",
  GroupLabel2 = "SM",
  Plot = FALSE,
  PlotOffset = FALSE,
  CheckESS = TRUE,
  MinESS = 100
)
```

```{r droplet-tested-boxplot, fig.cap="Boxplot of log over-dispersion parameters in pre-somitic and somitic mesoderm cells, respectively."}
DropletTableDisp <- as.data.frame(
  Droplet_Test_logFC0,
  Parameter = "Disp",
  Filter = FALSE
)
not_excluded <- DropletTableDisp$ResultDiffDisp != "ExcludedFromTesting"
for_plot <- DropletTableDisp[not_excluded, c("Disp1", "Disp2")]
for_plot <- reshape2::melt(for_plot)
ggplot(for_plot, aes(variable, log(value))) +
  geom_violin() +
  geom_boxplot(width = 0.25) +
  scale_x_discrete(labels = c("PSM", "SM")) +
  ylab("log(delta)") + xlab("")
```

### Global changes in variability

With this analysis, we do not detect global changes in expression variability.
We next profile changes in mean expression and expression variability on a 
gene-specific level.
For this, we use a log~2~ fold change threshold of 1 for mean expression testing 
and the default threshold of $\psi_0\approx0.41$ for differential variability 
testing.

```{r droplet-gene-specific-testing, fig.cap="Fold changes of average expression in presomitic mesoderm cells relative to somitic mesoderm cells are plotted again mean expression. Colour indicates genes that were excluded from differential expression test, and those with significantly higher mean expression in either group."}
Droplet_Test <- BASiCS_TestDE(
  Chain1 = PSM_MCMC,
  Chain2 = SM_MCMC,
  EpsilonM = 1,
  GroupLabel1 = "PSM",
  GroupLabel2 = "SM",
  Plot = FALSE,
  PlotOffset = FALSE,
  CheckESS = TRUE,
  MinESS = 100
)

BASiCS_PlotDE(Droplet_Test, Parameter = "Mean", Plot = "MA")
Droplet_TableMean <- as.data.frame(
  Droplet_Test,
  Parameter = "Mean",
  Filter = FALSE
)
```

```{r droplet-gene-specific-testing-resdisp, fig.cap="Differences in residual over-dispersion parameters in presomitic mesoderm cells relative to somitic mesoderm cells are plotted again mean expression. Colour indicates genes that were excluded from differential expression test, and those with significantly higher residual over-dispersion in either group."}
Droplet_TableResDisp <- as.data.frame(
  Droplet_Test,
  Parameter = "ResDisp",
  Filter = FALSE
)
BASiCS_PlotDE(Droplet_Test, Parameter = "ResDisp", Plot = "MA")
```

### Differential mean expression

We can now list the genes that were detected as differentially expressed and 
differentially variable ordered by their difference in mean 
expression/variability.
We first focus on genes that are differentially expressed between the two cell 
types.

```{r gene-lists-mean}
genenames <- readRDS("rds/genenames.rds")
## Highly expressed in somitic mesoderm
ind_sm <- Droplet_TableMean$ResultDiffMean == "SM+"
SM_mean <- Droplet_TableMean[ind_sm, ]
SM_mean <- SM_mean[order(SM_mean$MeanLog2FC, decreasing = FALSE), ]
SM_mean$Symbol <- genenames[SM_mean$GeneName, 2]

## Highly expressed in pre-somitic mesoderm
ind_psm <- Droplet_TableMean$ResultDiffMean == "PSM+"
PSM_mean <- Droplet_TableMean[ind_psm, ]
PSM_mean <- PSM_mean[order(PSM_mean$MeanLog2FC, decreasing = TRUE), ]
PSM_mean$Symbol <- genenames[PSM_mean$GeneName, 2]
```
