---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Nils Eling
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
    email: eling@ebi.ac.uk
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
  - name: John C. Marioni
    affiliation: 
    - *EBI
    - *CRUK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems, such as immunity and development. Single-cell RNA 
  sequencing is a powerful tool to quantify this heterogeneity, but it is prone 
  to strong technical noise. In this article, we describe a step-by-step 
  computational workflow which uses the BASiCS Bioconductor package to robustly 
  quantify expression variability within and between known groups of cells (such 
  as experimental conditions or cell types). BASiCS uses an integrated framework 
  for data normalisation, technical noise quantification and downstream 
  analyses, whilst propagating statistical uncertainty across these steps. 
  Within a single seemingly homogeneous cell population, BASiCS can identify 
  highly variable genes that exhibit strong heterogeneity as well as lowly 
  variable genes with stable expression. BASiCS also uses a probabilistic 
  decision rule to identify changes in expression variability between cell 
  populations, whilst avoiding confounding effects related to differences in 
  technical noise or in overall abundance. Using two publicly available 
  datasets, we guide users through a complete pipeline which includes 
  preliminary steps for quality control as well as data exploration 
  using the scater and scran Bioconductor packages. Data for the first case 
  study was generated using the Fluidigm\@ C1 system, in which extrinsic 
  spike-in RNA molecules were added as a control. The second dataset was 
  generated using a droplet-based system, for which spike-in RNA is not 
  available. This analysis provides an example, in which differential 
  variability testing reveals insights regarding a possible early cell fate 
  commitment process. The workflow is accompanied by a Docker image that 
  ensures the reproducibility of our results. 
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article: default
  BiocStyle::html_document: default
---


```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
# Use fig.width = 7 for html and fig.width = 6 for pdf
fig.width <- ifelse(on.bioc, 7, 6)
knitr::opts_chunk$set(
  cache = TRUE, warning = FALSE, message = FALSE, error = FALSE,
  cache.path = "cache/", fig.path = "figure/", fig.width = fig.width
)
```

```{r}
library(SingleCellExperiment)
library(scater)
library(scran)
library(BASiCS)
```

# C1 Fluidigm data: Analysis of naive CD4^+^ T cells {#Tcells}

For the first case study, we will use scRNA-seq data of CD4^+^ T cells, which 
were processed using the C1 Single-Cell Auto Prep System (Fluidigm^®^) using 
10–17 $\mu{}m$ integrated fluidic circuits (IFCs). 
Martinez-Jimenez _et al._ profiled naive and activated CD4^+^ T cells from 
young and old animals across two mouse strains to test for changes in 
expression variability that occur during organismal ageing 
[@Martinez-jimenez2017]. They extracted naive or effector memory CD4^+^ T cells
from spleens of young or old animals and filtered using either 
magnetic-activated cell sorting (MACS) or fluorescence activated cell sorting 
(FACS) (labelled as MACS-purified Naive, FACS-purified Naive or FACS-purified 
Effector Memory).
For clarification, naive CD4^+^ T cells are also referred to as 'unstimulated' 
CD4^+^ T cells.

In addition to profiling naive CD4^+^ T cells, the authors stimulated half 
of the naive cells for 3 hours using _in vitro_ antibody stimulation (labelled 
as Active).
They processed naive as well as activated CD4^+^ T cells using the 
C1 Fluidigm^®^ system to capture and lyse cells, and to reverse transcribe 
and amplify mRNA prior to sequencing.
The authors isolated cells from B6 (C57BL/6J, _Mus musculus domesticus_) and
CAST (_Mus musculus castaneus_) animals to profile the evolutionary conservation
of transcriptional dynamics during ageing.
Additionally, the authors added external spike-in RNA to aid in quantifying
technical variability across all cells. They performed all experiments in
replicates (also referred to as batches) to control for batch effects.

We will begin the workflow with obtaining the data before quality control, 
running the BASiCS model, and performing further downstream analysis.


## Obtaining the data

The raw counts of the full dataset can be obtained from ArrayExpress under the 
accession number 
[E-MTAB-4888](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-4888/).
In this dataset, the column names contain the library identifier of the original 
experiment, while the row names of the matrix store gene names. 
The dataset contains reads mapped to ERCC spike-in genes [@Rna2005], which 
`BASiCS` uses to estimate and remove technical noise.

```{r naive-data}
if (!file.exists("downloads/raw_data.txt")) {
  # Download raw counts file
  website <- "https://www.ebi.ac.uk/"
  folder <- "arrayexpress/files/E-MTAB-4888/"
  file <- "E-MTAB-4888.processed.1.zip"
  destfile <- "downloads/raw_data.txt.zip"
  download.file(
    paste0(website, folder, file),
    destfile = destfile
  )
  unzip("downloads/raw_data.txt.zip", exdir = "downloads")
  file.remove("downloads/raw_data.txt.zip")
}

# Read in raw data
CD4_raw <- read.table("downloads/raw_data.txt", header = TRUE, sep = "\t")
CD4_raw <- as.matrix(CD4_raw)

# Show row and column names
head(colnames(CD4_raw))
head(rownames(CD4_raw))

# ERCC spike-in genes
head(rownames(CD4_raw)[grepl("ERCC", rownames(CD4_raw))])
```

To link the library identifiers to the experimental conditions, the authors 
provide a metadata file that can be accessed online under the same accession 
number.

```{r selecting-serum-cells}
download_file <- function(file, website, folder, destfile = file) {
  if (!file.exists(destfile)) {
    download.file(paste0(website, folder, file), destfile)
  }
}

if (!file.exists("downloads/metadata_file.txt")) {
  # Download raw counts file
  website <- "https://www.ebi.ac.uk/"
  folder <- "arrayexpress/files/E-MTAB-4888/"
  file <- "E-MTAB-4888.additional.1.zip"
  destfile <- "downloads/metadata.txt.zip"
  download.file(
    paste0(website, folder, file),
    destfile = destfile
  )
  unzip("downloads/metadata.txt.zip", exdir = "downloads")
  file.remove("downloads/metadata.txt.zip")
}
# Read in metadata file
CD4_metadata <- read.table("downloads/metadata_file.txt", header = TRUE, sep = "\t")

# Save library identifier as rownames
rownames(CD4_metadata) <- CD4_metadata$X

# Show metadata entries
head(CD4_metadata)
```

This metadata file contains the library identifiers (*X*), the strain 
information (*Strain*), the relative age of the animals (*Age*), the 
stimulation state of the cells (*Stimulus*), the batch information indicating 
the different mice used (*Individuals*), and the purification method used 
(*Celltype*). 


## Filtering the data

In the next step, we will first generate a `r Biocpkg("SingleCellExperiment")` 
(SCE) object that contains all cells and that stores cell- and gene-specific 
metadata.
This data class offers convenient ways to subset, set and retrieve cell- 
and gene-specific information. We will also store information on the spike-in
molecules using the `altExp` 
accessor. This allows us to store alternative experimental assays other
than endogenous genes --- in this case, technical (spike-in) genes.

```{r CD4-SCE-object}
# Load library
bio_counts <- CD4_raw[!grepl("ERCC", rownames(CD4_raw)), ]
spike_counts <- CD4_raw[grepl("ERCC", rownames(CD4_raw)), ]
# Generate the SingleCellExperiment object
sce_CD4_all <- SingleCellExperiment(
  assays = list(counts = as.matrix(bio_counts)),
  colData = CD4_metadata[colnames(CD4_raw), ]
)

altExp(sce_CD4_all, "spike-ins") <- SummarizedExperiment(
  assays = list(counts = spike_counts)
)
```

For further downstream analysis, we select naive and activated CD4^+^ T cells 
from young B6 animals that were obtained using the MACS-based cell selection.
It is crucial that proper quality control (QC) and filtering is performed 
before running BASiCS, and, in general, before performing any computational
analysis on biological data.

```{r naive-activated-CD4-SCE-object}
ind_select <- sce_CD4_all$Strain == "Mus musculus domesticus" &
  sce_CD4_all$Age == "Young" &
  sce_CD4_all$Celltype == "MACS-purified Naive"
sce_naive_active <- sce_CD4_all[, ind_select]
```

Prior to cell-specific quality control, we will remove all genes that are not
detected in at least 2 cells across both conditions.
Furthermore, we remove genes that are not expressed with an average count of 1 
across all cells.
These thresholds need to be set specifically for each dataset, and careful
gene-specific quality metrics need to be closely examined as suggested by the 
`r Biocpkg("SimpleSingleCell")` Bioconductor workflow [@Lun2016].

```{r gene-selection}
# Remove genes
ind_expressed <- rowSums(counts(sce_naive_active) > 0) > 1 &
  rowMeans(counts(sce_naive_active)) >= 1
sce_naive_active <- sce_naive_active[ind_expressed, ]
```

In the next step, we will use the `r Biocpkg("scran")` and 
`r Biocpkg("scater")` Bioconductor packages for initial normalisation and 
visualisation of quality metrics [@Lun2016,@McCarthy2017].
The normalisation at this point is needed to avoid biases in the visualisation 
due to differences in the mRNA content between cells.
In line with the normalisation strategy of BASiCS, we use the spike-in reads 
for normalisation.

```{r pre-normalisation}
# Pre-normalisation of data for visualisation purposes
sce_naive_active <- computeSpikeFactors(sce_naive_active, spikes = "spike-ins")
sce_naive_active <- logNormCounts(sce_naive_active)
```

The `Biocpkg("SimpleSingleCell")` Bioconductor workflow and
the `OSCA` online book each provide an extensive 
overview on important aspects of how to perform low-level analysis of scRNA-seq 
data, including quality control [@Lun2016;@Amezquita2019].
The `calculateQCMetrics` function of the `scater` package can be used to 
calculate a number of cell- and gene-specific quality metrics.
Furthermore, `scater` offers the `runPCA` function to perform principal 
component analysis (PCA) across all cells.

```{r scater-QC-metrics}
# Calculate quality metrics
sce_naive_active <- addPerCellQC(sce_naive_active, use_altexps = TRUE)
sce_naive_active <- addPerFeatureQC(sce_naive_active)

clean_colnames <- make.names(colnames(colData(sce_naive_active)))
colnames(colData(sce_naive_active)) <- clean_colnames

# Calculate PCA
sce_naive_active <- scater::runPCA(sce_naive_active)
```

For the CD4^+^ T cell dataset, the authors removed empty capture sites and 
sites containing multiple cells or debris, as observed by visual inspection. 
Furthermore, they removed cells which had fewer than 1,000,000 total reads, 
cells where less than 20% of reads mapped to endogenous genes, cells in which 
less than 1,250 or more than 3,000 genes were detected, and cells with more than 
10% or less than 0.5% reads mapping to mitochondrial genes 
[@Martinez-jimenez2017].

Since these data have been quality filtered in the original publication, we will 
only visualise the distribution of quality metrics across the cells.
The highlighted quality metrics can otherwise be used to identify low quality 
cells. These metrics include: number of endogeneous genes 
detected per cell, total number of reads per cell, the fraction of reads mapping
to spike-in and endogeneous genes.
The main aim is to identify broken or dying cells, cells for which a very
small quantity of mRNA was captured, and possibly empty wells.
For an extenive discussion on quality metrics for scRNA-seq data, see Ilicic 
*et al.* [@Ilicic2016].

The activation status and batch information for each individual cell of the 
selected dataset can be seen in Figure \@ref(fig:visualisation-stimulus-batch).

```{r visualisation-stimulus-batch,fig.height=2, fig.cap="Visualisation of the cell stimulus (left) and the batch information per cell (right). Cells from two animals were captured either in the naive or activated state.", fig.pos='h'}
# Visualise the conditions and batch structure
p_stimulus <- scater::plotPCA(sce_naive_active, colour_by = "Stimulus")
p_batch <- scater::plotPCA(sce_naive_active, colour_by = "Individuals")
multiplot(p_stimulus, p_batch, cols = 2)
```

The cells split into two conditions: naive and activated CD4^+^ T cells.
Furthermore, we detect neither obvious batch effects nor outlier cells.

<!-- What is a doublet? -->
In the next step, we will visualise selected cell-specific quality metrics 
overlayed on the principal components.
Figure \@ref(fig:no-genes_total-counts) highlights the number of endogeneous 
genes detected per cell and the total number of counts.
Both of these metrics can be used to detect empty wells or broken cells 
(low values) and possible doublets (high values).

```{r no-genes-total-counts, fig.height=2, fig.cap="Visualisation of the number of biological genes detected per cell (left) and the total number of reads per cell (right).", fig.pos='h'}
# Visualise number of endogeneous genes detected
p_total_features <- scater::plotPCA(
    sce_naive_active, 
    colour_by = "detected"
  ) +
  scale_fill_viridis_c(name = "Number of genes", trans = "log10")

# Visualise log10-transformed total number of counts
p_total_counts <- scater::plotPCA(
    sce_naive_active, 
    colour_by = "sum"
  ) +
  scale_fill_viridis_c(name = "Number of counts", trans = "log10")
multiplot(p_total_features, p_total_counts, cols = 2)
```

We detect a higher number of expressed genes and a higher total read count in
activated cells. Furthermore, cells within each group show a wide distribution
of these quality measures, without clear outlying cells.

A widely used quality visualisation is to plot the total number (or fraction) of 
spike-in counts versus the total number (or fraction) of endogeneous counts 
(Figure \@ref(fig:ERCC-endogeneous)).
In such a plot, low quality wells are characterised by a high fraction of 
spike-in counts and a low fraction of endogeneous counts.

```{r ERCC-endogeneous, fig.height=3, fig.width=3, fig.cap="The number of spike-in counts per cells is plotted against the number of endogeneous counts per cell. A dashed line indicates the minimum level of spike-in counts, below which cells are removed.", fig.pos='h'}
# Plot the fraction of reads mapping to spike-ins
# versus the number of reads mapping to endogeneous genes
ggplot(as.data.frame(colData(sce_naive_active))) +
  geom_point(aes(sum, altexps_spike.ins_sum)) +
  geom_vline(xintercept = 10^5.8, linetype = "dashed", colour = "grey60") +
  scale_x_log10() +
  scale_y_log10()
```

Since we are lacking the information of the total reads per cell (including the 
non-mapped and intronic reads), we can solely filter cells based on the total 
number of endogeneous counts (assuming equal numbers of spike-in molecules per 
well). Nevertheless, we remove two cells that showed unusually low number of 
reads mapping to endogeneous genes. We indicate the threshold used for filtering
on the plot using a dashed line.

```{r}
# Remove two cells that appear to be outliers
ind_retain <- log10(sce_naive_active$sum) > 5.8
sce_naive_active <- sce_naive_active[, ind_retain]
```

It is crucial that an in-depth quality control is performed to remove 
low-quality or outlying cells. Furthermore, these quality control steps can
help users to understand possible confounding effects that may influence
the interpretation of downstream analyses.


## Calculating the molecule count for spike-in genes

BASiCS requires the absolute molecule count of the spike-in transcripts that
were added to each well. To calculate the molecule count, we require the 
dilution and the concentration of the spike-in mix.
For this, a table of the spike-in concentrations can be downloaded from 
[https://www.thermofisher.com](https://www.thermofisher.com/order/catalog/product/4456740).
The file contains the concentrations of 2 ERCC spike-in mixes.

For the CD4^+^ T cell data, the authors added a 1:50,000 dilution of the ERCC 
spike-in mix 1 [@Martinez-jimenez2017].
We will first download the concentration information.

```{r spike-in_download}
# Read in the spike-in concentrations
website <- "https://assets.thermofisher.com/"
folder <- "TFS-Assets/LSG/manuals/"
file <- "cms_095046.txt"
download_file(file, website, folder, "downloads/spike_info.txt")

ERCC_conc <- read.table(
  "downloads/spike_info.txt",
  sep = "\t", header = TRUE
)
```

The concentration is given in units of $aM\mu{}l^{-1}$. 
We will first calculate the concentration in $M\mu{}l^{-1}$.

```{r}
# Moles per micro litre
ERCC_mmul <- ERCC_conc$concentration.in.Mix.1..attomoles.ul. * (10^(-18))
```

From this, we can calculate the molecule count per $\mu{}l$ using the fact that
1 mole is comprised of $6.02214076 \times 10^{23}$ molecules.

```{r}
# Molecule count per micro litre
ERCC_countmul <- ERCC_mmul * (6.02214076 * (10^23))
```

During the preparation of the reaction mix, the authors diluted this mix by a
factor of 1:50,000 [@Martinez-jimenez2017].
The actual molecule number of spike-ins can therefore be calculated per $\mu{}l$:

```{r}
ERCC_count <- ERCC_countmul / 50000
```

The volume per well in the IFC chip is $9nl$
[https://www.fluidigm.com/faq/ifc-9](https://www.fluidigm.com/faq/ifc-9).
We therfore need to calculate the number of molecules in $9nl$ reaction mix.

```{r}
ERCC_count_final <- ERCC_count * 0.009
```

Depending on the technology used to capture and process single cells, as well 
as the dilution of the spike-in mix, the absolute number of spike-ins per 
reaction can change. The absolute amount of spike-ins per reaction is only used 
to calibrate the scale of the capture efficiency normalisation parameter 
$\nu_j$. Since the true concentration of each spike-in molecule is known, 
BASiCS uses the observed counts of each molecule for each cell to estimate the 
capture efficiency for each cell, $\nu_j$. Differences in the 
global scale across all wells of the level of spike-ins molecules thus do not 
affect the cell-to-cell ratios of $\nu_j$.
BASiCS assumes that the quantity of spike-ins is consistent in each well in
each experiment. While the quantity used must remain constant between wells and
experiments, the scale does not affect the results, provided the spike-ins
are at a reasonable level. In particular, they should not be in such a low 
concentration that they are rarely detected, and they should not be at such
a high concentration that the majority of the sequencing reads map to the 
spike-in molecules.

We can now use the molecule count to prepare the BASiCS data object.
To incorporate the spike-in molecule counts within the 
`r Biocpkg("SingleCellExperiment")` object that BASiCS requires, the first 
column must contain the names associated to the spike-in genes. 
The second column must contain the input number of molecules for the spike-in 
genes (amount per reaction).


```{r spike-info}
# Prepare the data.frame
ERCC_count <- data.frame(
  row.names = ERCC_conc$ERCC.ID,
  Names = ERCC_conc$ERCC.ID,
  count = ERCC_count_final
)
```
