\documentclass[9pt,a4paper,]{extarticle}

\usepackage{f1000_styles}

\usepackage[pdfborder={0 0 0}]{hyperref}

\usepackage[numbers]{natbib}
\bibliographystyle{unsrtnat}


%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother


% disable code chunks background
%\renewenvironment{Shaded}{}{}

% disable section numbers
\setcounter{secnumdepth}{0}

%% added by MLS, this is not in the F1000 style by default %%

\hypersetup{unicode=true,
            pdftitle={BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data},
            pdfkeywords={Single-cell RNA sequencing, expression variability, transcriptional noise, differential expression testing},
            colorlinks=true,
            linkcolor=Maroon,
            citecolor=Blue,
            urlcolor=Orange,
            breaklinks=true}

%% End added by MLS %%

\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}



\begin{document}
\pagestyle{front}

\title{BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data}

\author[1]{Alan O'Callaghan\thanks{\ttfamily a.b.o'callaghan@sms.ed.ac.uk}}
\author[2]{Nils Eling}
\author[3,4]{John C. Marioni}
\author[1,5]{Catalina A. Vallejos\thanks{\ttfamily catalina.vallejos@igmm.ed.ac.uk}}
\affil[1]{MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, EH4 2XU, UK}
\affil[2]{Department of Quantitative Biomedicine, University of Zurich, Winterthurerstrasse 190, CH-8057, Zurich, Switzerland}
\affil[3]{European Molecular Biology Laboratory, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK}
\affil[4]{Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Cambridge, CB2 0RE, UK}
\affil[5]{The Alan Turing Institute, British Library, 96 Euston Road, London, NW1 2DB, UK}

\maketitle
\thispagestyle{front}

\begin{abstract}
Cell-to-cell gene expression variability is an inherent feature of complex
biological systems, such as immunity and development. Single-cell RNA
sequencing is a powerful tool to quantify this heterogeneity, but it is prone
to strong technical noise. In this article, we describe a step-by-step
computational workflow that uses the BASiCS Bioconductor package to robustly
quantify expression variability within and between known groups of cells (such
as experimental conditions or cell types). BASiCS uses an integrated framework
for data normalisation, technical noise quantification and downstream
analyses, whilst propagating statistical uncertainty across these steps.
Within a single seemingly homogeneous cell population, BASiCS can identify
highly variable genes that exhibit strong heterogeneity as well as lowly
variable genes with stable expression. BASiCS also uses a probabilistic
decision rule to identify changes in expression variability between cell
populations, whilst avoiding confounding effects related to differences in
technical noise or in overall abundance. Using a publicly available
dataset, we guide users through a complete pipeline that includes
preliminary steps for quality control, as well as data exploration
using the scater and scran Bioconductor packages. The workflow is accompanied
by a Docker image that ensures the reproducibility of our results.
\end{abstract}

\section*{Keywords}
Single-cell RNA sequencing, expression variability, transcriptional noise, differential expression testing


\clearpage
\pagestyle{main}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

Parameter estimation is implemented in the \texttt{BASiCS\_MCMC} function using an
adaptive Metropolis within Gibbs algorithm \citep{Roberts2009}.
The primary inputs for \texttt{BASiCS\_MCMC} correspond to:

\begin{itemize}
\item
  \texttt{Data}: a \texttt{SingleCellExperiment} object created as described in the
  previous sections.
\item
  \texttt{N}: the total number of MCMC iterations.
\item
  \texttt{Thin}: thining period for output storage
  (only the \texttt{Thin}-th MCMC draw is stored).
\item
  \texttt{Burn}: the initial number of MCMC iterations to be discarded.
\item
  \texttt{Regression}: if \texttt{TRUE} a join prior is assigned to \(\mu_i\) and \(\delta_i\)
  \citep{Eling2018}, and residual over-dispersion values \(\epsilon_i\) are inferred.
  Alternatively, independent log-normal priors are assigned to \(\mu_i\) and
  \(\delta_i\) \citep{Vallejos2016}.
\item
  \texttt{WithSpikes}: if \texttt{TRUE} information from spike-in molecules is used to aid
  data normalisation and to quantify technical noise.
\item
  \texttt{PriorParam}: Defines the prior hyper-parameters to be used by
  \emph{\href{https://bioconductor.org/packages/3.12/BASiCS}{BASiCS}}. We recommend to use the \texttt{BASiCS\_PriorParam} function
  for this purpose. If \texttt{PriorMu\ =\ "EmpiricalBayes"} in \texttt{BASiCS\_PriorParam},
  \(\mu_i\)'s are assigned a log-normal prior with gene-specific location
  hyper-parameters defined via an
  empirical Bayes framework. Alternatively, if \texttt{PriorMu\ =\ "default"}, location
  hyper-parameter are set to be equal 0 for all genes.
\end{itemize}

{\small\bibliography{Workflow.bib}}

\end{document}
