\documentclass[9pt,a4paper,]{extarticle}

\usepackage{f1000_styles}

\usepackage[pdfborder={0 0 0}]{hyperref}

\usepackage[numbers]{natbib}
\bibliographystyle{unsrtnat}


%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother


% disable code chunks background
%\renewenvironment{Shaded}{}{}

% disable section numbers
\setcounter{secnumdepth}{0}

%% added by MLS, this is not in the F1000 style by default %%

\hypersetup{unicode=true,
            pdftitle={BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data},
            pdfkeywords={Single-cell RNA sequencing, expression variability, transcriptional noise, differential expression testing},
            colorlinks=true,
            linkcolor=Maroon,
            citecolor=Blue,
            urlcolor=Orange,
            breaklinks=true}

%% End added by MLS %%

\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}



\begin{document}
\pagestyle{front}

\title{BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data}

\author[1,2]{Nils Eling\thanks{\ttfamily eling@ebi.ac.uk}}
\author[3]{Alan O'Callaghan}
\author[1,2]{John C. Marioni}
\author[3,4]{Catalina A. Vallejos\thanks{\ttfamily catalina.vallejos@igmm.ed.ac.uk}}
\affil[1]{European Molecular Biology Laboratory, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK}
\affil[2]{Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Cambridge, CB2 0RE, UK}
\affil[3]{MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, EH4 2XU, UK}
\affil[4]{The Alan Turing Institute, British Library, 96 Euston Road, London, NW1 2DB, UK}

\maketitle
\thispagestyle{front}

\begin{abstract}
Cell-to-cell gene expression variability is an inherent feature of complex
biological systems, such as immunity and development. Single-cell RNA
sequencing is a powerful tool to quantify this heterogeneity, but it is prone
to strong technical noise. In this article, we describe a step-by-step
computational workflow which uses the BASiCS Bioconductor package to robustly
quantify expression variability within and between known groups of cells (such
as experimental conditions or cell types). BASiCS uses an integrated framework
for data normalisation, technical noise quantification and downstream
analyses, whilst propagating statistical uncertainty across these steps.
Within a single seemingly homogeneous cell population, BASiCS can identify
highly variable genes that exhibit strong heterogeneity as well as lowly
variable genes with stable expression. BASiCS also uses a probabilistic
decision rule to identify changes in expression variability between cell
populations, whilst avoiding confounding effects related to differences in
technical noise or in overall abundance. Using two publicly available
datasets, we guide users through a complete pipeline which includes
preliminary steps for quality control as well as data exploration
using the scater and scran Bioconductor packages. Data for the first case
study was generated using the Fluidigm@ C1 system, in which extrinsic
spike-in RNA molecules were added as a control. The second dataset was
generated using a droplet-based system, for which spike-in RNA is not
available. This analysis provides an example, in which differential
variability testing reveals insights regarding a possible early cell fate
commitment process. The workflow is accompanied by a Docker image that
ensures the reproducibility of our results.
\end{abstract}

\section*{Keywords}
Single-cell RNA sequencing, expression variability, transcriptional noise, differential expression testing


\clearpage
\pagestyle{main}

\hypertarget{reproducibility}{%
\subsection{Reproducibility}\label{reproducibility}}

The following software versions were used throughout this workflow:

\begin{itemize}
\item
  \textbf{R version}: R Under development (unstable) (2020-01-28 r77738)
\item
  \textbf{Bioconductor version}: 3.11
\item
  \textbf{R packages}: BASiCS 1.99.1, scran
  1.15.29, scater 1.15.32. Version
  numbers for all remaining packages is available at \protect\hyperlink{session-info}{Session Info}.
\end{itemize}

A Docker image containing all software requirements is available at {[}ADD LINK{]}.

\hypertarget{Tcells}{%
\section{\texorpdfstring{C1 Fluidigm data: Analysis of naive CD4\textsuperscript{+} T cells}{C1 Fluidigm data: Analysis of naive CD4+ T cells}}\label{Tcells}}

For the first case study, we will use scRNA-seq data of CD4\textsuperscript{+} T cells, which
were processed using the C1 Single-Cell Auto Prep System (Fluidigm\textsuperscript{®}) using
10--17 \(\mu{}m\) integrated fluidic circuits (IFCs).
Martinez-Jimenez \emph{et al.} profiled naive and activated CD4\textsuperscript{+} T cells from
young and old animals across two mouse strains to test for changes in
expression variability that occur during organismal ageing
\citep{Martinez-jimenez2017}. They extracted naive or effector memory CD4\textsuperscript{+} T cells
from spleens of young or old animals and filtered using either
magnetic-activated cell sorting (MACS) or fluorescence activated cell sorting
(FACS) (labelled as MACS-purified Naive, FACS-purified Naive or FACS-purified
Effector Memory).
For clarification, naive CD4\textsuperscript{+} T cells are also referred to as `unstimulated'
CD4\textsuperscript{+} T cells.

In addition to profiling naive CD4\textsuperscript{+} T cells, the authors stimulated half
of the naive cells for 3 hours using \emph{in vitro} antibody stimulation (labelled
as Active).
They processed naive as well as activated CD4\textsuperscript{+} T cells using the
C1 Fluidigm\textsuperscript{®} system to capture and lyse cells, and to reverse transcribe
and amplify mRNA prior to sequencing.
The authors isolated cells from B6 (C57BL/6J, \emph{Mus musculus domesticus}) and
CAST (\emph{Mus musculus castaneus}) animals to profile the evolutionary conservation
of transcriptional dynamics during ageing.
Additionally, the authors added external spike-in RNA to aid in quantifying
technical variability across all cells. They performed all experiments in
replicates (also referred to as batches) to control for batch effects.

We will begin the workflow with obtaining the data before quality control,
running the BASiCS model, and performing further downstream analysis.

\hypertarget{obtaining-the-data}{%
\subsection{Obtaining the data}\label{obtaining-the-data}}

The raw counts of the full dataset can be obtained from ArrayExpress under the
accession number
\href{https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-4888/}{E-MTAB-4888}.
In this dataset, the column names contain the library identifier of the original
experiment, while the row names of the matrix store gene names.
The dataset contains reads mapped to ERCC spike-in genes \citep{Rna2005}, which
\texttt{BASiCS} uses to estimate and remove technical noise.

{\small\bibliography{Workflow.bib}}

\end{document}
