---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Nils Eling
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
    email: eling@ebi.ac.uk
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
  - name: John C. Marioni
    affiliation: 
    - *EBI
    - *CRUK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems, such as immunity and development. Single-cell RNA 
  sequencing is a powerful tool to quantify this heterogeneity, but it is prone 
  to strong technical noise. In this article, we describe a step-by-step 
  computational workflow which uses the BASiCS Bioconductor package to robustly 
  quantify expression variability within and between known groups of cells (such 
  as experimental conditions or cell types). BASiCS uses an integrated framework 
  for data normalisation, technical noise quantification and downstream 
  analyses, whilst propagating statistical uncertainty across these steps. 
  Within a single seemingly homogeneous cell population, BASiCS can identify 
  highly variable genes that exhibit strong heterogeneity as well as lowly 
  variable genes with stable expression. BASiCS also uses a probabilistic 
  decision rule to identify changes in expression variability between cell 
  populations, whilst avoiding confounding effects related to differences in 
  technical noise or in overall abundance. Using two publicly available 
  datasets, we guide users through a complete pipeline which includes 
  preliminary steps for quality control as well as data exploration 
  using the scater and scran Bioconductor packages. Data for the first case 
  study was generated using the Fluidigm\@ C1 system, in which extrinsic 
  spike-in RNA molecules were added as a control. The second dataset was 
  generated using a droplet-based system, for which spike-in RNA is not 
  available. This analysis provides an example, in which differential 
  variability testing reveals insights regarding a possible early cell fate 
  commitment process. The workflow is accompanied by a Docker image that 
  ensures the reproducibility of our results. 
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article: default
  BiocStyle::html_document: default
---


```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
# Use fig.width = 7 for html and fig.width = 6 for pdf
fig.width <- ifelse(on.bioc, 7, 6)
knitr::opts_chunk$set(
  cache = TRUE, warning = FALSE, message = FALSE, error = FALSE,
  cache.path = "cache/", fig.path = "figure/", fig.width = fig.width
)
```

<!-- I simplified this section to avoid too much mathematical formulation.
Readers can go to the original papers for more details.
I think it is important to emphasise the intution/interpretation instead. -->
## Quantifying cell-to-cell transcriptional variability - `BASiCS`

The `r Biocpkg("BASiCS")` package implements a Bayesian hierarchical framework 
which borrows information across all genes and cells to robustly quantify 
transcriptional variability [@Vallejos2015BASiCS].
Similar to the approach adopted in `r Biocpkg("scran")`, `r Biocpkg("BASiCS")` 
infers cell-specific global scaling normalisation parameters. 
However, instead of inferring these as a pre-processing step, 
`r Biocpkg("BASiCS")` uses an integrated approach in which data normalisation 
and downstream analyses are performed simultaneously --- whilst propagating 
statistical uncertainty. 
To quantify technical noise, the original implementation of 
`r Biocpkg("BASiCS")` uses information from extrinsic spike-in molecules as 
control features, but the model has been extended to address situations in which 
spike-ins are not available [@Eling2018]. 

`r Biocpkg("BASiCS")` summarises the distribution of gene expression through 
gene-specific *mean* ($\mu_i$) and *over-dispersion* ($\delta_i$) parameters. 
Mean parameters $\mu_i$ quantify the overall expression for each gene $i$
across the population of cells under study.
Instead, $\delta_i$ captures the excess of variability that is observed with 
respect to what would be expected in a homogeneous cell population, after 
taking into account technical noise. 
This is used as a proxy to quantify transcriptional variability.
Moreover, to account for the strong association that is typically observed 
between mean expression and over-dispersion estimates, we recently introduced 
gene-specific *residual over-dispersion* parameters $\epsilon_i$ [@Eling2018]. 
Similar to DM values implemented in `r Biocpkg("scran")`, these are defined as 
deviations with respect to an overall regression trend that captures the 
relationship between mean and over-dispersion values.

Parameter estimation is implemented in the `BASiCS_MCMC` function, which can be 
run using four different major settings (Table 1). 
If spike-in molecules are available and `WithSpikes = TRUE` (default), the model 
uses this information in order to infer technical noise.
Alternatively, if `WithSpikes = FALSE`, the approach introduced in [@Eling2018]
is applied.
We recommend to use `Regression = TRUE` (default) as the underlying informative
prior improves the stability of posterior estimates for small sample sizes and 
lowly expressed genes [@Eling2018]. 
Moreover, it provides additional flexibility for downstream analysis as it also
infers gene-specific residual over-dispersion parameters $\epsilon_i$. 

<!---
If the regression between over-dispersion and mean expression should be 
performed, the `Regression` parameters is set to `TRUE` (default).
If the user decides to set `Regression = FALSE`, `BASiCS` will not estimate 
the regression trend, and will not supply the residual over-dispersion 
parameters $\epsilon_i$. ---> 

: Four settings available for the the `BASiCS_MCMC` function.

                        No regression         Regression           
----------------------- --------------------  -------------------
Using spike-in reads    `WithSpikes = TRUE`   `WithSpikes = TRUE`  
                        `Regression = FALSE`  `Regression = TRUE` 
No spike-ins available  `WithSpikes = FALSE`  `WithSpikes = FALSE` 
                        `Regression = FALSE`  `Regression = TRUE`
----------------------- --------------------  -------------------

<!-- Paragraph to be edited --->
The algorithm implemented in the `BASiCS_MCMC` function is an adaptive 
Metropolis within Gibbs sampler [@Roberts2009].
The `BASiCS_MCMC` function returns a `BASiCS_Chain` object, which can be used 
for further downstream analyses, many of which are detailed in this workflow.
These objects contain draws from Markov chain Monte Carlo (MCMC) samplers,
which are used to infer the posterior distribution over the model parameters
[@Smith1993].
Briefly, the posterior distribution quantifies how probable different parameter 
values are given the observed data. However, before assessing the posterior
distribution, we must first ensure that the MCMC sampler has converged to
its stationary distribution, and has sampled efficiently from this distribution
[@Cowles1996]. If these conditions are not met, then the estimated parameters
may be inaccurate. The `r CRANpkg("coda")` CRAN package contains a variety of 
functions to assess the convergence of a sampled MCMC chain. 
To use `coda` functions, the individual chains returned by `BASiCS` need to be 
transformed into a MCMC object that `coda` recognises using the `coda::mcmc` 
function. `BASiCS` also offers a number of functions to visualise and assess the 
convergence of MCMC chains. In particular, we will use 
`BASiCS_EffectiveSize` and `BASiCS_DiagPlot` to calculate and visualise the
effective sample size generated by the MCMC samplers.

[talk about downstream analyses; hvg/lvg; differential testing 
then mention the ability to perform differential testing;
finally the extention to account for mean/over-dispersion]

## Other steps [title tbc]

The `r Biocpkg("goseq")` Bioconductor package offers functions to detect the 
enrichment of gene ontologies (GOs) among user-specified gene sets [@Young2010].
Furthermore, `goseq` corrects for gene length biases, which is useful for full 
length scRNA-seq data as highlighted in the first section. 
In this workflow, we will use `goseq` to detect GO enrichment among 
differentially expressed sets of genes.

For downstream analysis, such as GO enrichment analysis or the biological 
interpretation of individual genes, we need to (i) link each gene's ID to its 
symbol and (ii) calculate each gene's length.
For the first task, the `r Biocpkg("biomaRt")` Bioconductor package annotates a 
wide range of gene and gene product identifiers [@Durinck2005] by accessing the 
BioMart software suite ([http://www.biomart.org](http://www.biomart.org)).
We can use `biomaRt` to link the **Mus musculus** gene IDs and to their gene 
symbols (also referred to as 'gene name'): 





