---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Nils Eling
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
    email: eling@ebi.ac.uk
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
  - name: John C. Marioni
    affiliation: 
    - *EBI
    - *CRUK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems, such as immunity and development. Single-cell RNA 
  sequencing is a powerful tool to quantify this heterogeneity, but it is prone 
  to strong technical noise. In this article, we describe a step-by-step 
  computational workflow which uses the BASiCS Bioconductor package to robustly 
  quantify expression variability within and between known groups of cells (such 
  as experimental conditions or cell types). BASiCS uses an integrated framework 
  for data normalisation, technical noise quantification and downstream 
  analyses, whilst propagating statistical uncertainty across these steps. 
  Within a single seemingly homogeneous cell population, BASiCS can identify 
  highly variable genes that exhibit strong heterogeneity as well as lowly 
  variable genes with stable expression. BASiCS also uses a probabilistic 
  decision rule to identify changes in expression variability between cell 
  populations, whilst avoiding confounding effects related to differences in 
  technical noise or in overall abundance. Using two publicly available 
  datasets, we guide users through a complete pipeline which includes 
  preliminary steps for quality control as well as data exploration 
  using the scater and scran Bioconductor packages. Data for the first case 
  study was generated using the Fluidigm\@ C1 system, in which extrinsic 
  spike-in RNA molecules were added as a control. The second dataset was 
  generated using a droplet-based system, for which spike-in RNA is not 
  available. This analysis provides an example, in which differential 
  variability testing reveals insights regarding a possible early cell fate 
  commitment process. The workflow is accompanied by a Docker image that 
  ensures the reproducibility of our results. 
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article: default
  BiocStyle::html_document: default
---


```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
# Use fig.width = 7 for html and fig.width = 6 for pdf
fig.width <- ifelse(on.bioc, 7, 6)
knitr::opts_chunk$set(
  cache = TRUE, warning = FALSE, message = FALSE, error = FALSE,
  cache.path = "cache/", fig.path = "figure/", fig.width = fig.width
)
```

# Introduction

<!--- scRNA-seq and the different types of heterogeneity ---> 
<!--- Nils to revisit and add additional references if required --->
Single-cell RNA-sequencing (scRNA-seq) enables the study of genome-wide 
transcriptional heterogeneity in cell populations that remains otherwise 
undetected in bulk experiments [@Stegle2015; @Prakadan2017; @Patange2018]. 
<!--- Applications of scRNA-seq range from characterising cell types in immunity 
[@Lonnberg2017;  @Villani2017; @Zheng2017] and development [@Ibarra-Soria2018; 
@Wagner2018; @Pijuan-Sala2019] to dissecting the mechanisms for cell fate 
commitment [@Goolam2016; @Ohnishi2014]. ---> 
On the broadest level, this heterogeneity can reflect the presence of distinct 
cell subtypes or states. <!---, which could be characterised through 
clustering [@Kiselev2019]. --->
Alternatively, cell-to-cell expression heterogeneity can be due to gradual 
changes along processes that evolve over time, such as development and differentiation. <!--- and that can be characterised 
using pseudotime inference methods [@Saelens2019]. --->
Several clustering and pseudotime inference methods have been developed to
characterise these sources of heterogeneity [@Kiselev2019; @Saelens2019].
However, there is a limited availability of computational tools tailored 
to study more subtle variability within a seemingly homogeneous cell population. 
This variability can be due to deterministic or stochastic events that regulate gene expression and, among others, has been reported to increase prior to cell phate decisions [@Mojtahedi2016] as well as throughout ageing [@Martinez-jimenez2017]. 

<!--- Describes aims for the workflow and introduces BASiCS ---> 
This article complements existing scRNA-seq workflows based on the
Bioconductor package ecosystem (e.g. [@Lun2016; @Kim2019]). 
We describe a step-by-step analysis which uses `r Biocpkg("scater")` and 
`r Biocpkg("scran")` to perform quality control (QC) as well as 
initial exploratory analyses [@McCarthy2017; @Lun2016]. 
To robustly quantify transcriptional variability within and
between pre-specified cell populations (such as experimental conditions or
cell types) we use `r Biocpkg("BASiCS")` [@Vallejos2015; @Vallejos2016;
@Eling2017] --- a Bayesian hierarchical framework that simultaneously performs 
data normalisation (global scaling), technical noise quantification and selected 
downstream analyses, whilst propagating statistical uncertainty across these 
steps. Among others, `r Biocpkg("BASiCS")`, has led to new insights about the 
heterogeneity of immune cells [@Martinez-jimenez2017].

<!--- BASiCS downstream analyses within a single population --->
Within a population of cells, `r Biocpkg("BASiCS")` decomposes the total 
observed variability in expression measurements into technical and biological components [@Vallejos2015]. 
This enables the identification of *highly variable genes* (HVGs) that capture
the major sources of heterogeneity within the analysed cells [@Brennecke2013]. 
HVG detection is often used as feature selection, to identify the input 
set of genes for subsequent analyses. 
`r Biocpkg("BASiCS")` can also highlight *lowly variable genes* (LVGs) that 
exhibit stable expression across the population of cells.
These may relate to essential cellular functions and can assist the development
of new data normalisation or integration strategies [@Lin2019]. 

<!--- BASiCS downstream analyses between populations --->
In order to compare expression patterns between two or more pre-specified 
groups of cells, `r Biocpkg("BASiCS")` provides a probabilistic decision rule to 
perform differential expression analyses [@Vallejos2016; @Eling2018].
Whilst several differential expression tools have been proposed for scRNA-seq 
data (e.g. [@Kharchenko2014; @Finak2015]), some evidence suggests that 
these do not generally outperform popular bulk RNA-seq tools [@Soneson2018]. 
Moreover, most of these methods are only designed to uncover changes in overall
expression, ignoring the more complex patterns that can arise at the single cell 
level [@Lahnemann2020]. 
Instead, `r Biocpkg("BASiCS")` embraces the high granularity of scRNA-seq data,
uncovering changes in cell-to-cell expression variability that are not 
confounded by differences in technical noise or in overall expression. 

<!--- TBC description of BASiCS downstream analysis, with focus on differential --->
<!--- add brief description for LVG/HVG 
<!--- Differential analyses ---> 
<!--- add reference to 11 challenges paper, flexible changes in expression ---> 
<!-- DE testing isn't always between (2) conditions (blocking, continuous, etc) -->
<!--- Since the era of RNA sequencing, methods for differential expression testing 
of transcript counts across conditions have been developed 
[@Anders2010; @Robinson2009]. 
Due to high technical variability and sparsity in scRNA-seq data, new 
approaches were developed for differential expression testing 
for scRNA-seq data [@Katayama2013; @Kharchenko2014; @Delmans2016]. 
In contrast to bulk samples, scRNA-seq measures variations in gene expression 
across a population of cells, and can therefore be used to test for changes in 
expression variability between two conditions. 
To do this, `BASiCS` compares the gene-specific over-dispersion parameters 
between two conditions. These parameters are independent of technical noise 
and can be used as proxy for biological variability [@Vallejos2016]. 
Similar to the mean-variability trend observed for normalised scRNA-seq data 
[@Brennecke2013], the estimates for over-dispersion parameters decrease with 
mean expression [@Vallejos2016].
To correct for this, BASiCS has been extended to model the mean-variability 
relationship and capture residual over-dispersion estimates that show no 
association to mean expression. 
Therefore, this extension allows to test changes in mean expression in parallel 
to changes in variability [@Eling2018].  ---> 
<!--- TBC Outline - wider use of bioc libraries + case studies ---> 
<!---
Two case studies exemplify the use of `BASiCS` for non-UMI and UMI 
scRNA-seq data. In the first case, `BASiCS` can be used to detect highly and 
lowly variable genes and to obtain robust, gene-specific estimates to assess
biological variability in naive CD4^+^ T cells 
[@Martinez-jimenez2017]; for a similar workflow see [@Kim2019].
Furthermore, we compare naive to activated CD4^+^ T cells to highlight the use 
of `BASiCS` to test for changes in mean expression and expression variability.
In the second case, we use droplet-based scRNA-seq data to detect more subtle 
transcriptional changes during embryonic somitogenesis [@Ibarra-Soria2018].
--->

<!--- Brief outline + reproducibility ---> 
In this manuscript, we briefly discuss the sources of variability that arise in 
scRNA-seq data and some of the strategies that have been designed to control or 
attenuate technical noise in these assays. 
We also summarise the main features of the Bioconductor packages that are used
throughout this workflow and provide a description for the underlying statistical 
model implemented in `r Biocpkg("BASiCS")`.
This includes practical guidance to assess the convergence of the Markov Chain 
Monte Carlo (MCMC) algorithm that is used to infer model parameters as well as 
recommendations to interpret and to post-process the model outputs. 
Finally, we provide two step-by-step case studies using naive CD4^+^ T cells 
[@Martinez-jimenez2017] and samples collected during embryonic somitogenesis 
[@Ibarra-Soria2018]. These examples were chosen to illustate the usage of 
`r Biocpkg("BASiCS")` in the context of different scRNA-seq protocols. 

All source code used to generate the results presented in this article is 
available at https://github.com/VallejosGroup/BASiCSWorkflow. To ensure the 
reproducibility of this workflow, the analysis environment and all software 
dependencies are provided as a Docker [@Boettiger2014] image at: [ADD LINK].


<!--- 

## Outline

This article is organised as follows:

1. [Transcriptional noise](#noise-in-scRNA-seq-datasets): [TBC - summary about transcriptional noise]

2. [Methods](#methods): We describe the main Bioconductor packages used in 
  this worflow and introduce the different settings implemented in the 
  `r Biocpkg("BASiCS")` Bioconductor package. 
  
3. [Reproducibility](#reproducibility): [DESCRIBE COMP REQUIREMENTS, DOCKER, 
ETC --- CREATE SECTION IN MANUSCRIPT!]

4. [C1 Fluidigm data](#Tcells): We present an end-to-end analysis workflow to 
  obtain, process, quality filter raw expression counts of CD4^+^ T cells and how to 
  perform analysis using `r Biocpkg("BASiCS")`. 
  This section is sub-divided into a [single group case](#Tcells-single), 
  where `BASiCS` is used to normalise data and to detect highly and lowly 
  variable genes, and a [two group case](#Tcells-two), for which we perform 
  differential mean expression and differential variability testing.

5. [10X Genomics data](#droplet-data): We highlight the use of `BASiCS` for 
  scRNA-seq data that do not contain technical spike-in transcript. 
  In this case, technical noise is estimated via experimental replication.
  Similar to the CD4^+^ T cells case, we perform differential mean expression 
  and differential variability testing to detect early cell fate decision events 
  in mammalian somitogenesis. 
  
---> 

## Sources of variability in scRNA-seq datasets

<!--- text to be revisited ---> 
<!--- An overarching goal of scRNA-seq is to characterise gene expression 
heterogeneity with cellular resolution. ---> 
The focus of this article is to use scRNA-seq data in order to quantify the
strength of cell-to-cell expression heterogeneity within seemingly homogeneous 
cell populations. 
Here, we briefly describe the underlying sources of heterogeneity that can be 
captured by cell-to-cell variability estimates derived from scRNA-seq data. 

Stochastic variability within a cell population --- often referred to as
transcriptional *noise* --- can arise from intrinsic and extrinsic sources
[@Elowitz2002; @Eling2019]. 
Classically, extrinsic noise is defined as stochastic fluctuations induced by 
different dynamic cellular states (e.g. cell cycle, metabolism, intra- and
inter-cellular signalling) 
[@Zopf2013; @Iwamoto2016; @Kiviet2014]. 
In contrast, intrinsic noise arises from stochastic effects on biochemical 
processes such as transcription and translation [@Elowitz2002].
Intrinsic noise can be modulated by genetic and epigenetic modifications (such 
as mutations, histone modifications, CpG island length and nucleosome 
positioning) [@Eberwine2015; @Faure2017; @Morgan2018] and is usually measured 
at the gene level [@Elowitz2002]. 
Cell-to-cell gene expression variability estimates derived from scRNA-seq data 
capture a combination of these effects, as well as deterministic regulatory 
mechanisms [@Eling2019]. 
Moreover, these variability estimates can also be inflated by the technical 
noise that is typically observed in scRNA-seq data [@Brennecke2013].

<!--- Experimental strategies to tackle technical noise --->
Different strategies have been incorporated into scRNA-seq protocols to control 
or attenuate technical noise. 
For example, external RNA spike-in molecules (such as the set introduced by the 
External RNA Controls Consortium, ERCC [@Rna2005]) can be added to each cell’s 
lysate in a (theoretically) known fixed quantity.
Spike-ins can assist quality control steps [@McCarthy2017], data normalisation
[@Vallejos2017] and can be used to infer technical noise [@Brennecke2013].
Another strategy is to tag individual cDNA molecules using unique molecular 
identifiers (UMIs) before PCR amplification [@Islam2014]. 
Reads that contain the same UMI can be collapsed into a single molecule count,
attenuating technical variability associated to cell-to-cell differences
in amplification and sequencing depth (these technical biases are not fully 
removed unless sequencing to saturation [@Vallejos2017]). 
However, despite the benefits associated to the use of spike-ins and UMIs, 
these are not available for all scRNA-seq protocols [@Haque2017]. 

# Methods {#methods}

This step-by-step scRNA-seq workflow is primarily based on the Bioconductor 
package ecosystem [@Amezquita2019]. 
A graphical overview is provided in Figure \@ref(fig:overview) 
and its main components are described below. 

```{r overview, out.width='50%', fig.cap = 'Graphical overview for the scRNA-seq analysis workflow described in this manuscript. Starting from a SingleCellExperiment object, we use the scater and scran Bioconductor packages to perform quality control and initial exploratory analyses. The primary focus of this workflow is to robustly quantify transcriptional heterogeneity within seemingly homogeneous cell populations. For this purpose, we apply the BASiCS Bioconductor package, illustrating how BASiCS can be used to analyse a single or multiple pre-specified groups of cells.', echo=FALSE, fig.pos='h'}
knitr::include_graphics("figure/Overview.png")
```

## Input data - `SingleCellExperiment`

Here, we use the `r Biocpkg("SingleCellExperiment")` package to convert an input 
matrix of raw read-counts (molecule counts for UMI-based protocols) into a 
`SingleCellExperiment` object which can store scRNA-seq data and its associated metadata, such as gene- and cell-specific information. 
Moreover, when available, the same object can be used to store read-counts for 
technical spike-in molecules (these can be accessed via the `altExp()` method).
A major advantage of using a `SingleCellExperiment` object as the input for 
scRNA-seq analyses is that it enables interoperability across a large number of 
Bioconductor packages [@Amezquita2019]. 

## Quality control and exploratory analyses - `scater` and `scran`

An critical step in scRNA-seq analyses is to apply QC diagnostics, removing low
quality samples (wells or droplets, depending on the protocol) that may distort
downstream analyses. 
Among others, QC can help to identify samples that contain broken cells, that 
are empty or that contain multiple cells [@Ilicic2016classification]. 
Moreover, lowly expressed genes for which less reliable information is 
available are typically also removed. 
To perform QC of scRNA-seq data, we use the `r Biocpkg("scater")` Bioconductor 
package [@McCarthy2017].
In `r Biocpkg("scater")`, the `calculateQCMetrics` function can be used to 
calculate standard QC metrics for each cell (e.g. percentage of mitochondrial 
reads) and gene (e.g. percentage of zeroes across all cells). 
The package also provides a suite of visualisation tools that can be used to 
explore the data under study and its associated QC diagnostic metrics. 

The `r Biocpkg("scran")` package offers additional tools for QC 
diagnostics and a variety of functions scRNA-seq data analysis [@Lun2016].
It can perform *global scaling* data normalisation, calculating cell-specific scaling factors that capture global differences in read-counts across cells (e.g. due to sequencing depth and PCR amplification) [@Lun2016pooling].
In `r Biocpkg("scran")`, users can also explore how the observed variability in
expression counts can be decomposed into technical and biological 
sources (see `decomposeVar`). 
Moreover, the `trendVar` function can be used to infer an overall trend between 
gene-specific mean and variance estimates. 
To derive gene-specific variability estimates that are not confounded by this 
overall trend, the `DM` function calculates the distance between gene-specific 
squared coefficients of variation (CV$^2$) and a rolling median along the range
of mean expression values [@Kolodziejczyk2015cell].
DM estimates enable exploratory analyses of cell-to-cell heterogeneity, but a measure of uncertainty is not readily available. As such, gene-specific 
downstream inference (e.g. differential variability testing) is precluded.

<!-- I simplified this section to avoid too much mathematical formulation.
Readers can go to the original papers for more details.
I think it is important to emphasise the intution/interpretation instead. -->
## Analysis of cell-to-cell transcriptional variability - `BASiCS`

The `r Biocpkg("BASiCS")` package uses a Bayesian hierarchical framework 
which borrows information across all genes and cells to robustly quantify 
transcriptional variability [@Vallejos2015BASiCS].
Similar to the approach adopted in `r Biocpkg("scran")`, `r Biocpkg("BASiCS")` 
infers cell-specific global scaling normalisation parameters. 
However, instead of inferring these as a pre-processing step, 
`r Biocpkg("BASiCS")` uses an integrated approach in which data normalisation 
and downstream analyses are performed simultaneously, thereby propagating
statistical uncertainty. 
To quantify technical noise, the original implementation of 
`r Biocpkg("BASiCS")` uses information from extrinsic spike-in molecules as 
control features, but the model has been extended to address situations in which 
spike-ins are not available [@Eling2018].

`r Biocpkg("BASiCS")` summarises expression patterns through 
gene-specific *mean* ($\mu_i$) and *over-dispersion* ($\delta_i$) parameters. 
Mean parameters $\mu_i$ quantify the overall expression for each gene $i$
across the population of cells under study.
Instead, $\delta_i$ captures the excess of variability that is observed with 
respect to what would be expected in a homogeneous cell population, after 
taking into account technical noise. 
This is used as a proxy to quantify transcriptional variability.
Moreover, to account for the strong association that is typically observed 
between mean expression and over-dispersion estimates, we recently introduced 
gene-specific *residual over-dispersion* parameters $\epsilon_i$ [@Eling2018]. 
Similar to DM values implemented in `r Biocpkg("scran")`, these are defined as 
deviations with respect to an overall regression trend that captures the 
relationship between mean and over-dispersion values.

Parameter estimates are obtained via the `BASiCS_MCMC` function, which can be 
run using four different major settings (Table 1). 
`WithSpikes` indicates whether spike-in molecules are used to infer technical
noise (default: `WithSpikes = TRUE`). 
We recommend to use `Regression = TRUE` (default) as the underlying prior 
improves the stability of posterior estimates for small sample sizes and 
lowly expressed genes [@Eling2018]. 
Moreover, it provides additional flexibility for downstream analysis, inferring
gene-specific residual over-dispersion parameters $\epsilon_i$ that are not 
confounded by changes in mean expression.

<!---
If the regression between over-dispersion and mean expression should be 
performed, the `Regression` parameters is set to `TRUE` (default).
If the user decides to set `Regression = FALSE`, `BASiCS` will not estimate 
the regression trend, and will not supply the residual over-dispersion 
parameters $\epsilon_i$. ---> 

: Four settings available for the the `BASiCS_MCMC` function.

                        No regression         Regression           
----------------------- --------------------  -------------------
Using spike-in reads    `WithSpikes = TRUE`   `WithSpikes = TRUE`  
                        `Regression = FALSE`  `Regression = TRUE` 
No spike-ins available  `WithSpikes = FALSE`  `WithSpikes = FALSE` 
                        `Regression = FALSE`  `Regression = TRUE`
----------------------- --------------------  -------------------

<!-- Paragraph to be edited --->
An adaptive Metropolis within Gibbs sampler [@Roberts2009] algorithm is 
implemented in `BASiCS_MCMC`. 
The total number of iterations for the algorithm is indicated by `N`. 
The `BASiCS_MCMC` function returns a `BASiCS_Chain` object, which can be used 
for further downstream analyses.
The output object contains 

These objects contain draws from Markov chain Monte Carlo (MCMC) samplers,
which are used to infer the posterior distribution over the model parameters
[@Smith1993].
Briefly, the posterior distribution quantifies how probable different parameter 
values are given the observed data. However, before assessing the posterior
distribution, we must first ensure that the MCMC sampler has converged to
its stationary distribution, and has sampled efficiently from this distribution
[@Cowles1996]. If these conditions are not met, then the estimated parameters
may be inaccurate. The `r CRANpkg("coda")` CRAN package contains a variety of 
functions to assess the convergence of a sampled MCMC chain. 
To use `coda` functions, the individual chains returned by `BASiCS` need to be 
transformed into a MCMC object that `coda` recognises using the `coda::mcmc` 
function. `BASiCS` also offers a number of functions to visualise and assess the 
convergence of MCMC chains. In particular, we will use 
`BASiCS_EffectiveSize` and `BASiCS_DiagPlot` to calculate and visualise the
effective sample size generated by the MCMC samplers.

[talk about downstream analyses; hvg/lvg; differential testing 
then mention the ability to perform differential testing;
finally the extention to account for mean/over-dispersion]

