---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
    email: "a.b.o'callaghan@sms.ed.ac.uk"
  - name: Nils Eling
    affiliation: 
    - &UZH Department of Quantitative Biomedicine, University of Zurich,
      Winterthurerstrasse 190, CH-8057, Zurich, Switzerland
  - name: John C. Marioni
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems, such as immunity and development. Single-cell RNA 
  sequencing is a powerful tool to quantify this heterogeneity, but it is prone 
  to strong technical noise. In this article, we describe a step-by-step 
  computational workflow that uses the BASiCS Bioconductor package to robustly 
  quantify expression variability within and between known groups of cells (such 
  as experimental conditions or cell types). BASiCS uses an integrated framework 
  for data normalisation, technical noise quantification and downstream 
  analyses, whilst propagating statistical uncertainty across these steps. 
  Within a single seemingly homogeneous cell population, BASiCS can identify 
  highly variable genes that exhibit strong heterogeneity as well as lowly 
  variable genes with stable expression. BASiCS also uses a probabilistic 
  decision rule to identify changes in expression variability between cell 
  populations, whilst avoiding confounding effects related to differences in 
  technical noise or in overall abundance. Using a publicly available 
  dataset, we guide users through a complete pipeline that includes 
  preliminary steps for quality control, as well as data exploration 
  using the scater and scran Bioconductor packages. The workflow is accompanied 
  by a Docker image that ensures the reproducibility of our results. 
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article:
    fig_width: 6
    fig_height: 3.5
  BiocStyle::html_document:
    fig_width: 8
    fig_height: 4.5
---



```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
theme_set(theme_bw())
# Use fig.width = 7 for html and fig.width = 6 for pdf
# fig.width <- ifelse(on.bioc, 10, 6)
if (knitr::is_html_output()) {
  out_width <- "700px"
  out_height <- "600px"
} else if (knitr::is_latex_output()) {
  out_width <- "2.5in"
  out_height <- "3.5in"
}
knitr::opts_chunk$set(
  warning = FALSE, message = FALSE, error = FALSE,
  cache = 2, cache.path = "cache_main/",
  # fig.pos = "h",
  fig.path = "figure/"
)
```


# Introduction

<!--- scRNA-seq and the different types of heterogeneity ---> 
<!--- Nils to revisit and add additional references if required --->
Single-cell RNA-sequencing (scRNA-seq) enables the study of genome-wide 
transcriptional heterogeneity in cell populations that is not 
captured by bulk experiments [@Stegle2015; @Prakadan2017; @Patange2018]. 
On the broadest level, this heterogeneity can reflect the presence of distinct 
cell subtypes or states. 
Alternatively, it can be due to gradual changes along biological processes, 
such as development and differentiation. 
Several clustering and pseudotime inference methods have been developed to
characterise these types of heterogeneity [@Kiselev2019; @Saelens2019].
However, there is a limited availability of computational tools tailored 
to study more subtle variability within seemingly homogeneous cell populations. 
This variability can reflect deterministic or stochastic events that regulate
gene expression and, among others, has been reported to increase prior to cell
fate decisions [@Mojtahedi2016] as well as during ageing [@Martinez-jimenez2017]. 

<!--- Narrow focus to transcriptional noise --->
Stochastic variability within a seemingly homogeneous cell population --- often 
referred to as transcriptional *noise* --- can arise from intrinsic and 
extrinsic sources [@Elowitz2002; @Eling2019]. 
Extrinsic noise refers to stochastic fluctuations induced by 
different dynamic cellular states (e.g. cell cycle, metabolism, 
intra/inter-cellular signalling) [@Zopf2013; @Iwamoto2016; @Kiviet2014]. 
In contrast, intrinsic noise arises from stochastic effects on biochemical 
processes such as transcription and translation [@Elowitz2002].
Intrinsic noise can be modulated by genetic and epigenetic modifications (such 
as mutations, histone modifications, CpG island length and nucleosome 
positioning) [@Eberwine2015; @Faure2017; @Morgan2018] and usually occurs 
at the gene level [@Elowitz2002]. 
Cell-to-cell gene expression variability estimates derived from scRNA-seq data 
capture a combination of these effects, as well as deterministic regulatory 
mechanisms [@Eling2019]. 
Moreover, these variability estimates can also be inflated by the technical 
noise that is typically observed in scRNA-seq data [@Brennecke2013].

<!--- Experimental strategies to tackle technical noise --->
Different strategies have been incorporated into scRNA-seq protocols to control 
or attenuate technical noise. 
For example, external RNA spike-in molecules (such as the set introduced by the 
External RNA Controls Consortium, ERCC [@Rna2005]) can be added to each cellâ€™s 
lysate in a (theoretically) known fixed quantity.
Spike-ins can assist quality control steps [@McCarthy2017], data normalisation
[@Vallejos2017] and can be used to infer technical noise [@Brennecke2013].
Another strategy is to tag individual cDNA molecules using unique molecular 
identifiers (UMIs) before PCR amplification [@Islam2014]. 
Reads that contain the same UMI can be collapsed into a single molecule count,
attenuating technical variability associated to cell-to-cell differences
in amplification and sequencing depth (these technical biases are not fully 
removed unless sequencing to saturation [@Vallejos2017]). 
However, despite the benefits associated to the use of spike-ins and UMIs, 
these are not available for all scRNA-seq protocols [@Haque2017].

<!--- Introduce BASiCS --->
The Bioconductor package `r Biocpkg("BASiCS")` implements a Bayesian 
hierarchical framework that accounts for both technical and biological sources 
of noise in scRNA-seq datasets [@Vallejos2015; @Vallejos2016; @Eling2017]. 
`r Biocpkg("BASiCS")` jointly performs data normalisation, technical noise 
quantification and downstream analyses, whilst propagating statistical 
uncertainty across these steps. 
These features are implemented within a probabilistic model that builds upon a
negative binomial framework, a widely used distribution in the context of bulk 
and scRNA-seq experiments [@Svensson2020;@Townes2020;@Townes2019]. 
<!--- emphasis, this is critical ---> 
Critically, `r Biocpkg("BASiCS")` enables the quantification of transcriptional
variability within a population of cells, while accounting for the overall 
mean-variance relationship that typically arises in scRNA-seq data [@Eling2018].
Furthermore, when available, `r Biocpkg("BASiCS")` can also leverage extrinsic 
spike-in molecules to aid data normalisation. 

<!--- CV: removed comment about the no-spikes extension. I think it's implicit above ---> 
<!--- but a recent extension allows usage in the absence of 
spike-in molecules [@Eling2018]. ---> 

<!---
In particular, we
encourage the use of UMIs when performing a scRNAseq study.
It is well-established that scRNAseq data is distributed
according to a negative binomial distribution when UMIs are used in the 
sequencing protocol [@Svensson2020;@Townes2020;@Townes2019]. Thus, the
distributional assumptions of `r Biocpkg("BASiCS")` are more likely to be valid
for datasets wherein UMIs have been used. Furthermore, 
`r Biocpkg("BASiCS")` leverages spike-in molecules to aid in normalisation when
available, though a recent extension of the model allows the application of
the model in the absence of spike-in molecules [@Eling2018]. Moreover,
`r Biocpkg("BASiCS")` enables the quantification of variability within a 
population, while accounting for the overall mean-variance relationship in
an scRNAseq dataset [@Eling2018].
---> 

<!--- Describes aims for the workflow and introduces BASiCS ---> 
This article complements existing scRNA-seq workflows based on the Bioconductor 
ecosystem (e.g. [@Lun2016; @Kim2019]), providing a detailed framework for 
transcriptional variability analyses using `r Biocpkg("BASiCS")`. 
We describe a step-by-step workflow that uses `r Biocpkg("scater")` 
[@McCarthy2017] and `r Biocpkg("scran")` [@Lun2016] to perform quality control 
(QC) as well as initial exploratory analyses.
Our analysis pipeline includes practical guidance to assess the convergence of 
the Markov Chain Monte Carlo (MCMC) algorithm that is used to infer model 
parameters in `r Biocpkg("BASiCS")`, as well as recommendations to interpret 
and post-process the model outputs. 
Finally, through a case study in the context of immune cells, we illustrate
how `r Biocpkg("BASiCS")` can be used to identify highly and lowly variable
genes within a cell population, as well as to compare expression profiles
between experimental conditions or cell types. 

All source code used to generate the results presented in this article is 
available [on Github](https://github.com/VallejosGroup/BASiCSWorkflow).
To ensure the 
reproducibility of this workflow, the analysis environment and all software 
dependencies are provided as a Docker image [@Boettiger2015]. The image 
can be obtained from
[Docker Hub](https://hub.docker.com/repository/docker/alanocallaghan/bocker).


# Methods {#methods}

This step-by-step scRNA-seq workflow is primarily based on the Bioconductor 
package ecosystem [@Amezquita2019]. 
A graphical overview is provided in Figure \@ref(fig:overview) 
and its main components are described below. 

```{r overview, out.width=out_width, out.height=out_height, fig.cap = 'Graphical overview for the scRNA-seq analysis workflow described in this manuscript. Starting from a matrix of expression counts, we use the scater and scran Bioconductor packages to perform QC and initial exploratory analyses. To robustly quantify transcriptional heterogeneity within seemingly homogeneous cell populations, we apply the BASiCS Bioconductor package and  illustrate how BASiCS can be used to analyse a single or multiple pre-specified groups of cells.', echo=FALSE}
knitr::include_graphics("figure/Overview.png")
```


### Input data

```{r}
library("SingleCellExperiment")
```

We use `r Biocpkg("SingleCellExperiment")` to convert an input
matrix of raw read-counts (molecule counts for UMI-based protocols) into a 
`SingleCellExperiment` object that can also store its associated 
metadata, such as gene- and cell-specific information. 
Moreover, when available, the same object can also store read-counts for 
spike-in molecules (see `?altExp`).
A major advantage of using a `SingleCellExperiment` object as the input for 
scRNA-seq analyses is the interoperability across a large number of 
Bioconductor packages [@Amezquita2019]. 

## QC and exploratory data analysis

```{r}
library("scater")
library("scran")
```

An critical step in scRNA-seq analyses is QC, removing low quality samples that 
may distort downstream analyses. 
In this step, we use QC diagnostics to identify and remove samples that 
correspond to broken cells, that are empty, or that contain multiple cells 
[@Ilicic2016]. We  also typically remove lowly expressed genes that represent 
less reliable information. 
The [*OSCA*](https://osca.bioconductor.org/) online book provides an extensive
overview on important aspects of how to perform QC of scRNA-seq data, including
exploratory analyses [@Amezquita2019].

Here, we use the `r Biocpkg("scater")` package [@McCarthy2017] to calculate 
QC metrics for each cell (e.g. total read-count) and gene (e.g. percentage of 
zeroes across all cells), respectively. 
Moreover, we use the visualisation tools implemented in `r Biocpkg("scater")` to 
explore the input dataset and its associated QC diagnostic metrics. 
For further exploratory data analysis we use the `r Biocpkg("scran")` package 
[@Lun2016].
The latter can perform *global scaling* normalisation, calculating 
cell-specific scaling factors that capture global differences in read-counts 
across cells (e.g. due to sequencing depth and PCR amplification) 
[@Lun2016pooling]. 
To quantify transcriptional variability, `r Biocpkg("scran")` can be used to 
infer an overall trend between mean expression and the squared coefficent of 
variation (CV^2^) for each gene. 
To derive variability estimates that are not confounded by this overall trend, 
`r Biocpkg("scran")` also defines gene-specific DM (distance to the mean) 
estimates as the distance between CV$^2$ and a rolling median along the range 
of mean expression values [@Kolodziejczyk2015cell].
DM estimates enable exploratory analyses of cell-to-cell gene expression 
variability, but a measure of statistical uncertainty is not readily available 
for these estimates. 
As such, gene-specific downstream inference (such as differential variability 
testing) is precluded.


## BASiCS - Bayesian Analysis of Single Cell Sequencing data

```{r}
library("BASiCS")
```

The `r Biocpkg("BASiCS")` package uses a Bayesian hierarchical framework 
that borrows information across all genes and cells to robustly quantify 
transcriptional variability [@Vallejos2015BASiCS].
Similar to the approach adopted in `r Biocpkg("scran")`, `r Biocpkg("BASiCS")` 
infers cell-specific global scaling normalisation parameters. 
However, instead of inferring these as a pre-processing step, 
`r Biocpkg("BASiCS")` uses an integrated approach wherein data normalisation 
and downstream analyses are performed simultaneously, thereby propagating
statistical uncertainty. 
To quantify technical noise, the original implementation of 
`r Biocpkg("BASiCS")` uses information from extrinsic spike-in molecules as 
control features, but the model has been extended to address situations wherein
spike-ins are not available [@Eling2018].

`r Biocpkg("BASiCS")` summarises the expression pattern for each gene through 
gene-specific *mean* and *over-dispersion* parameters. 
Mean parameters $\mu_i$ quantify the overall expression for each gene $i$
across the cell population under study.
In contrast, $\delta_i$ captures the excess of variability that is observed with 
respect to what would be expected in a homogeneous cell population, beyond 
technical noise. 
`r Biocpkg("BASiCS")` uses $\delta_i$ as a proxy to quantify transcriptional 
variability.
To account for the strong relationship that is typically observed 
between gene-specific mean expression and over-dispersion estimates, 
Eling *et al.*  [@Eling2018] introduced a joint prior specification for 
these parameters. 
This joint prior formulation has been observed to improve posterior inference
when the data is less informative (e.g. small sample size, lowly expressed 
genes), borroring information across all genes to infer an overall trend that
captures the relationship between mean and over-dispersion.
The trend is then used to derive gene-specific *residual over-dispersion* 
parameters $\epsilon_i$ that are not confounded by mean expression. 
Similar to DM values implemented in `r Biocpkg("scran")`, these are defined as 
deviations with respect to the overall trend.

Within a population of cells, `r Biocpkg("BASiCS")` decomposes the total 
observed variability in expression measurements into technical and biological 
components [@Vallejos2015]. 
This enables the identification of *highly variable genes* (HVGs) that capture
the major sources of heterogeneity within the analysed cells [@Brennecke2013]. 
HVG detection is often used as feature selection, to identify the input 
set of genes for subsequent analyses. 
`r Biocpkg("BASiCS")` can also highlight *lowly variable genes* (LVGs) that 
exhibit stable expression across the population of cells.
These may relate to essential cellular functions and can assist the development
of new data normalisation or integration strategies [@Lin2019].

`r Biocpkg("BASiCS")` also provides a probabilistic decision rule to 
perform differential expression analyses between two (or more) pre-specified 
groups of cells [@Vallejos2016; @Eling2018].
While several differential expression tools have been proposed for scRNA-seq 
data (e.g. [@Kharchenko2014; @Finak2015]), some evidence suggests that 
these do not generally outperform popular bulk RNA-seq tools [@Soneson2018]. 
Moreover, most of these methods are only designed to uncover changes in overall
expression, ignoring the more complex patterns that can arise at the single cell 
level [@Lahnemann2020]. 
Instead, `r Biocpkg("BASiCS")` embraces the high granularity of scRNA-seq data,
uncovering changes in cell-to-cell expression variability that are not 
confounded by differences in technical noise or in overall expression.

# Case study: analysis of naive CD4^+^ T cells {#Tcells}

As a case study, we use scRNA-seq data generated for CD4^+^ T cells
using the C1 Single-Cell Auto Prep System (Fluidigm^Â®^). 
Martinez-Jimenez _et al._ profiled naive (hereafter also referred to as 
unstimulated) and activated (3 hours using _in vitro_ antibody stimulation) 
CD4^+^ T cells from young and old animals across two mouse strains to study 
changes in expression variability during ageing and upon immune activation 
[@Martinez-jimenez2017].
They extracted naive or effector memory CD4^+^ T cells from spleens of young or 
old animals, obtaining purified populations using either magnetic-activated cell 
sorting (MACS) or fluorescence activated cell sorting (FACS).
External ERCC spike-in RNA [@Rna2005] was added to aid the quantification of 
technical variability across all cells and all experiments were performed in 
replicates (hereafter also referred to as batches).

## Downloading the data

The matrix with raw read counts can be obtained from ArrayExpress under the 
accession number 
[E-MTAB-4888](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-4888/).
In the matrix, column names contain library identifiers and row names
display gene Ensembl identifiers. 

```{r naive-data}
if (!file.exists("downloads/"))
  dir.create("downloads", showWarnings = FALSE)
if (!file.exists("downloads/raw_data.txt")) {
  website <- "https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-4888/"
  file <- "E-MTAB-4888.processed.1.zip"
  download.file(
    paste0(website, file),
    destfile = "downloads/raw_data.txt.zip"
  )
  unzip("downloads/raw_data.txt.zip", exdir = "downloads")
  file.remove("downloads/raw_data.txt.zip")
}

CD4_raw <- read.table("downloads/raw_data.txt", header = TRUE, sep = "\t")
CD4_raw <- as.matrix(CD4_raw)
```

The input matrix contains data for `r format(ncol(CD4_raw), big.mark=",")`
cells and `r format(nrow(CD4_raw), big.mark=",")`
genes (including `r sum(grepl("ERCC", rownames(CD4_raw)))` ERCC spike-ins). 
Information about experimental conditions and other metadata is available 
under the same accession number. 

```{r selecting-serum-cells}
if (!file.exists("downloads/metadata_file.txt")) {
  website <- "https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-4888"
  file <- "E-MTAB-4888.additional.1.zip"
  download.file(
    paste0(website, file),
    destfile = "downloads/metadata.txt.zip"
  )
  unzip("downloads/metadata.txt.zip", exdir = "downloads")
  file.remove("downloads/metadata.txt.zip")
}

CD4_metadata <- read.table(
  "downloads/metadata_file.txt",
  header = TRUE,
  sep = "\t"
)

# Save sample identifiers as rownames
rownames(CD4_metadata) <- CD4_metadata$X
```

The columns in the metadata file contain library identifiers (`X`), strain 
information (`Strain`; *Mus musculus castaneus* or *Mus musculus domesticus*), 
the age of the animals (`Age`; young or old), stimulation state of the cells 
(`Stimulus`; naive or activated), batch information (`Individuals`; associated 
to different mice), and cell type information (`Celltype`; via FACS or MACS 
purification). 

Here, we convert the data and metadata described above into a 
`SingleCellExperiment` object. 
For this purpose, we first separate the input matrix of expression counts into
two matrices associated to intrinsic genes and external spike-ins, respectively. 
Within the `SingleCellExperiment` object, the latter is stored separately
as an *alternative experiment* (see `?altExp`).

```{r CD4-SCE-object}
# Separate intrinsic from ERCC counts
bio_counts <- CD4_raw[!grepl("ERCC", rownames(CD4_raw)), ]
spike_counts <- CD4_raw[grepl("ERCC", rownames(CD4_raw)), ]
# Generate the SingleCellExperiment object
sce_CD4_all <- SingleCellExperiment(
  assays = list(counts = as.matrix(bio_counts)),
  colData = CD4_metadata[colnames(CD4_raw), ]
)
# Add read-counts for spike-ins as an alternative experiment
altExp(sce_CD4_all, "spike-ins") <- SummarizedExperiment(
  assays = list(counts = spike_counts)
)
```

Hereafter, our analysis focuses on naive and activated CD4^+^ T cells obtained 
from young *Mus musculus domesticus* animals, purified using MACS-based cell sorting. 
Here, we extract these 
`r sum(sce_CD4_all$Strain == "Mus musculus domesticus" & sce_CD4_all$Age == "Young" & sce_CD4_all$Celltype == "MACS-purified Naive")` samples.

```{r naive-activated-CD4-SCE-object}
ind_select <- sce_CD4_all$Strain == "Mus musculus domesticus" &
  sce_CD4_all$Age == "Young" &
  sce_CD4_all$Celltype == "MACS-purified Naive"
sce_naive_active <- sce_CD4_all[, ind_select]
sce_naive_active
```

## Annotation

Input read counts were annotated using Ensembl gene identifiers. 
In order to facilitate the visualisation and interpretation of results, it is 
often useful to generate a mapping from Ensembl gene IDs to gene symbols using 
the BioMart software suite ([http://www.biomart.org](http://www.biomart.org)) 
via the Bioconductor package, `r Biocpkg("biomaRt")` [@Durinck2009].
These packages can also be used to obtain gene-pathways mappings and other
information such as gene length, useful for performing functional analysis
of the gene sets identified in downstream analyses. 

```{r obtain-gene-symbols}
if(!file.exists("rds/"))
  dir.create("rds", showWarnings = FALSE)

library(biomaRt)

if (!file.exists("rds/genenames.rds")) {
  # Initialize mart and dataset
  ensembl <- useMart(
    biomart = "ensembl",
    dataset = "mmusculus_gene_ensembl"
  )
  
  # Select gene ID and gene name
  genenames <- getBM(
    attributes = c("ensembl_gene_id", "external_gene_name"),
    mart = ensembl
  )

  rownames(genenames) <- genenames$ensembl_gene_id
  saveRDS(genenames, "rds/genenames.rds")
} else {
  genenames <- readRDS("rds/genenames.rds")
}
```

We add this information as `rowData` within the `SingleCellExperiment` 
object created above. 

```{r}
# Merge biomaRt annotation
my.rowdata <- data.frame(ensembl_gene_id = rownames(sce_naive_active))
my.rowdata <- merge(my.rowdata, genenames, by = "ensembl_gene_id", all.x = TRUE)
rownames(my.rowdata) <- rownames(sce_naive_active)
# Check to see that the order is correct after merge
# Sum must be equal to zero
sum(my.rowdata$ensembl_gene_id != rownames(my.rowdata))
# Add to the SingleCellExperiment object
rowData(sce_naive_active) <- my.rowdata
```

## QC and exploratory data analysis

The data available at 
[E-MTAB-4888](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-4888/) have
been already filtered to remove poor quality samples.
The QC applied in [@Martinez-jimenez2017] removed cells with: (i) fewer 
than 1,000,000 total reads, (ii) less than 20% of reads mapped to 
endogenous genes, (iii) less than 1,250 or more than 3,000 detected genes and 
(iv) more than 10% or fewer than 0.5% of reads mapped to mitochondrial genes.
As an illustration, we visualise some of these metrics. 
We also include another widely used QC diagnostic plot that compares the total 
number (or fraction) of spike-in counts versus the total number (or fraction) of 
endogeneous counts.
In such a plot, low quality samples are characterised by a high fraction of 
spike-in counts and a low fraction of endogeneous counts 
(see Figure \@ref(fig:PerCellQC)).

```{r PerCellQC, fig.cap="Cell-level QC metrics. The total number of endogenous read-counts (excludes non-mapped and intronic reads) is plotted against the total number of detected genes (left) and the total number of spike-in read-counts (right)."}
# Calculate and plot per cell QC metrics
sce_naive_active <- addPerCellQC(sce_naive_active, use_altexps = TRUE)
p_cellQC1 <- plotColData(
  sce_naive_active, 
  x = "sum", 
  y = "detected") +
  xlab("Total engogenous reads per cell") +
  ylab("Number of detected genes per cell") +
  theme(axis.text.x = element_text(hjust = 1, angle = 45))
p_cellQC2 <- plotColData(
  sce_naive_active, 
  x = "sum", 
  y = "altexps_spike-ins_sum") +
  xlab("Total engogenous reads per cell") +
  ylab("Total spike-in reads per cell") +
  theme(axis.text.x = element_text(hjust = 1, angle = 45))

library(patchwork)
p_cellQC1 + p_cellQC2
```

We can also visualise these metrics with respect to cell-level metadata, such
as the experimental conditions (active vs unstimulated) and the different mice 
that cells were collected from 
(see Figure \@ref(fig:experimental-condition-batch)). 

```{r experimental-condition-batch, fig.cap="Cell-level QC metrics according to cell-level metadata. The total number of endogenous reads (excludes non-mapped and intronic reads) is plotted against the total number of detected genes. Colour indicates the experimental condition (left) and animal of origin (right) for each cell."}
p_stimulus <- plotColData(
    sce_naive_active,
    x = "sum",
    y = "detected", 
    colour_by = "Stimulus"
  ) +
  xlab("Total engogenous reads per cell") +
  ylab("Number of detected genes per cell") +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
p_batch <- plotColData(
    sce_naive_active,
    x = "sum",
    y = "detected", 
    colour_by = "Individuals"
  ) +
  xlab("Total engogenous reads per cell") +
  ylab("Number of detected genes per cell") +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
p_stimulus + p_batch
```

To further explore the underlying structure of the data, we compute global
scaling normalisation factors using `r Biocpkg("scran")` and perform a 
principal component analysis (PCA) of log-transformed normalised expression 
counts using `r Biocpkg("scater")`.
As seen in Figure \@ref(fig:pca-visualisation-stimulus-batch), this analysis 
suggests the absence of strong batch effects.
It should be noted that `r Biocpkg("scran")` normalisation
is not strictly necessary in the `r Biocpkg("BASiCS")` workflow and we only
use it here as part of the exploratory data analysis.
Moreover, count-based models for dimensionality reduction (e.g. [@Townes2019;@Lopez2018]) could be used as an alternative to PCA, 
removing the need for log normalisation .

```{r pca-visualisation-stimulus-batch, fig.cap="First two principal components of log-transformed expression counts after scran normalisation. Colour indicates the experimental condition (left) and animal of origin (right) for each cell."}
# Global scaling normalisation + log tranformation + PCA
sce_naive_active <- computeSumFactors(sce_naive_active)
sce_naive_active <- logNormCounts(sce_naive_active)
sce_naive_active <- runPCA(sce_naive_active)
p_stimulus <- plotPCA(sce_naive_active, colour_by = "Stimulus") +
  theme(legend.position = "bottom")
p_batch <- plotPCA(sce_naive_active, colour_by = "Individuals") +
  theme(legend.position = "bottom")
p_stimulus + p_batch
```

In addition to cell-specific QC, we also recommend the use of a gene filtering 
step prior to the use of `r Biocpkg("BASiCS")`. 
The purpose of this filter is to remove lowly expressed genes that were largely
undetected through sequencing, making reliable variability estimates difficult
to obtain.
Here, we remove all genes that are not detected in at least 5 cells across both 
experimental conditions or that have an average read count below 1. 
These thresholds can vary across datasets and should be informed by 
gene-specific QC metrics such as those shown in
Figure \@ref(fig:gene-selection).

```{r gene-selection, fig.cap="Average read-count for each gene is plotted against the number of cells in which that gene was detected. Dashed grey lines are shown at the thresholds below which genes are removed."}
# Calculate per gene QC metrics
sce_naive_active <- addPerFeatureQC(sce_naive_active, exprs_values = "counts")
# Remove genes with zero total counts across all cells
sce_naive_active <- sce_naive_active[rowData(sce_naive_active)$detected != 0, ]
# Transform 'detected' metadata into number of cells 
rowData(sce_naive_active)$detected_cells <- 
  rowData(sce_naive_active)$detected * ncol(sce_naive_active) / 100
# Define inclusion criteria for genes
rowData(sce_naive_active)$include_gene <- rowData(sce_naive_active)$mean >= 1 &
  rowData(sce_naive_active)$detected_cells >= 5
plotRowData(
    sce_naive_active,
    x = "detected_cells",
    y = "mean",
    colour_by = "include_gene") +
  xlab("Total engogenous reads per cell") +
  ylab("Number of detected genes per cell") +
  scale_x_log10() +
  scale_y_log10() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
geom_vline(xintercept = 5, linetype = "dashed", col = "grey60") +
geom_hline(yintercept = 1, linetype = "dashed", col = "grey60")
# Apply gene filter
sce_naive_active <- sce_naive_active[rowData(sce_naive_active)$include_gene, ]
```

Subsequently, we also require users to remove spike-in molecules that were not 
captured through sequencing. We will do this separately for naive and active
cells separately.

```{r spike-ins-present}
ind_active <- sce_naive_active$Stimulus == "Active"
ind_naive <- sce_naive_active$Stimulus == "Unstimulated"
spikes <- assay(altExp(sce_naive_active))
detected_spikes_active <- rowSums(spikes[, ind_active] > 0) > 0
detected_spikes_naive <- rowSums(spikes[, ind_naive] > 0) > 0
detected_spikes <- detected_spikes_naive & detected_spikes_active
altExp(sce_naive_active) <- altExp(sce_naive_active)[detected_spikes, ]
```

The final dataset used in subsequent analyses contains 
`r ncol(sce_naive_active)` cells, `r nrow(sce_naive_active)` genes and 
`r nrow(altExp(sce_naive_active))` spike-ins.

## Input data for BASiCS

Here, we apply the `r Biocpkg("BASiCS")` model separately to cells from each
experimental condition (`r sum(sce_naive_active$Stimulus == "Unstimulated")` 
naive and `r sum(sce_naive_active$Stimulus == "Active")` activated cells). 
We create separate `SingleCellExperiment` objects for each group of cells. 

```{r SCE-separation}
sce_naive <- sce_naive_active[, ind_naive]
sce_active <- sce_naive_active[, ind_active]
```

`r Biocpkg("BASiCS")` requires the user to update these objects with additional
information, using a specific format.
Firstly, if multiple batches of sequenced cells are available (e.g. multiple 
donors that cells were extracted from, or multiple sequencing batches from the 
same experimental condition), this information must be included under the 
`BatchInfo` label as part of the cell-level metadata.

```{r BatchInfo}
colData(sce_naive)$BatchInfo <- colData(sce_naive)$Individuals
colData(sce_active)$BatchInfo <- colData(sce_active)$Individuals
```

If spike-ins will be used to aid data normalisation and technical noise 
quantification, `r Biocpkg("BASiCS")` also requires the number of spike-in 
molecules that were added to each well. 
For each spike-in gene $i$, this corresponds to:

$$ \mu_{i} = C_i \times 10^{-18} \times (6.022 \times 10^{23}) 
\times V \times D \hspace{0.5cm} \mbox{where,} $$

- $C_i$ is the concentration for the spike-in $i$ (measured in $aM\mu{}l^{-1}$),
- $V$ is the volume added into each well (measure in $nl$) and
- $D$ is a dilution factor. 

The remaining factors in the equation above are unit conversion constants
(e.g. from moles to molecules). 
For the CD4^+^ T cell data, the authors added a 1:50,000 dilution of the ERCC 
spike-in mix 1 and a volume of $9nl$ was added into each well (see [https://www.fluidigm.com/faq/ifc-9](https://www.fluidigm.com/faq/ifc-9)).
Finally, input concentrations $C_i$ can be downloaded from 
[https://assets.thermofisher.com/TFS-Assets/LSG/manuals](https://assets.thermofisher.com/TFS-Assets/LSG/manuals/cms_095046.txt).

```{r spike-in_download}
if (!file.exists("downloads/spike_info.txt")) {
  website <- "https://assets.thermofisher.com/TFS-Assets/LSG/manuals"
  file <- "cms_095046.txt"
  download.file(
    paste0(website, file),
    destfile = "downloads/spike_info.txt"
  )  
}

ERCC_conc <- read.table("downloads/spike_info.txt", sep = "\t", header = TRUE)
```

Based on this information, the calculation above proceeds as follows

```{r, ercc-mul}
# Moles per micro litre
ERCC_mmul <- ERCC_conc$concentration.in.Mix.1..attomoles.ul. * (10^(-18))
# Molecule count per micro litre (1 mole comprises 6.02214076 x 10^{23} molecules)
ERCC_countmul <- ERCC_mmul * (6.02214076 * (10^23))
# Application of the dilution factor (1:50,000)
ERCC_count <- ERCC_countmul / 50000
# Multiplying by the volume added into each well
ERCC_count_final <- ERCC_count * 0.009
```

To update the `sce_naive` and `sce_active` objects, the user must create a 
`data.frame` whose first column contains the labels associated to 
the spike-in molecule (e.g. ERCC-00130) and whose second column contains the 
input number of molecules calculated above.
We add this information as metadata for `altExp(sce_naive)` and 
`altExp(sce_active)`, respectively. 

```{r spike-info}
SpikeInput <- data.frame(
  Names = ERCC_conc$ERCC.ID,
  count = ERCC_count_final
)
# Exclude spike-ins not included in the input SingleCellExperiment objects
# and ensure the order of the rows is identical
SpikeInput <- SpikeInput[match(rownames(altExp(sce_naive)), SpikeInput$Names), ]

metadata(sce_naive)$SpikeInput <- SpikeInput
metadata(sce_active)$SpikeInput <- SpikeInput
```

## Parameter estimation using BASiCS

Parameter estimation is implemented in the `BASiCS_MCMC` function using an 
adaptive Metropolis within Gibbs algorithm [@Roberts2009].
The primary inputs for `BASiCS_MCMC` correspond to:

- `Data`: a `SingleCellExperiment` object created as described in the 
  previous sections. 
- `N`: the total number of MCMC iterations.
- `Thin`: thining period for output storage
  (only the `Thin`-th MCMC draw is stored).
- `Burn`: the initial number of MCMC iterations to be discarded.
- `Regression`: if `TRUE` a join prior is assigned to $\mu_i$ and $\delta_i$ 
  [@Eling2018], and residual over-dispersion values $\epsilon_i$ are inferred. 
  Alternatively, independent log-normal priors are assigned to $\mu_i$ and 
  $\delta_i$ [@Vallejos2016]. 
- `PriorMu`: If `"EmpiricalBayes"`, $\mu_i$'s are assigned a log-normal prior 
with gene-specific location hyper-parameters defined via an empirical Bayes 
framework. Alternatively, if `PriorMu = "default"`, location hyper-parameter are 
set to be equal 0 for all genes. 
- `WithSpikes`: if `TRUE` information from spike-in molecules is used to aid
  data normalisation and to quantify technical noise. 

As a default, we recommend to use `Regression = TRUE` as we have observed that
the joint prior introduced by Eling *et al.* leads to more stable, particularly 
for small sample sizes and lowly expressed genes. 
Moreover, the joint prior formulation enables users to obtain a measure of
transcriptional variability that is not confounded by mean expression. 
We also recommend to use `PriorMu = "EmpiricalBayes"` as we have observed 
that the empirical Bayes framework [REF] improves estimation performance for 
sparser datasets. 
Additional optional parameters can be used to store the generated output 
(`StoreChains`, `StoreDir`, `RunName`) and to monitor the progress of the 
algorithm (`PrintProgress`). 

Here, we run the MCMC sampler separately for naive and activated cells.
We use 40,000 iterations, discarding the initial 20,000 iterations.
We recommend this setting as a default choice, as we have observed it to 
ensure good convergence for the algorithm across multiple datasets.
However, for large datasets and less sparse datasets, a lower number of 
iterations may be sufficient. 
Practical guidance about the diagnostics criteria required to assess the 
performance of the MCMC algorithn is provided in the next section.

```{r MCMC-run, eval = FALSE}
chain_naive <- BASiCS_MCMC(
  Data = sce_naive,
  N = 40000,
  Thin = 20,
  Burn = 20000,
  Regression = TRUE,
  PriorMu = "EmpiricalBayes",
  WithSpikes = TRUE,
  StoreChains = TRUE,
  StoreDir = "rds/",
  RunName = "naive"
)

chain_active <- BASiCS_MCMC(
  Data = sce_active,
  N = 40000,
  Thin = 20,
  Burn = 20000,
  Regression = TRUE,
  PriorMu = "EmpiricalBayes",
  WithSpikes = TRUE,
  StoreChains = TRUE,
  StoreDir = "rds/",
  RunName = "active"
)
```

This first of these samplers takes 127 minutes to complete on a 3.4 GHz Intel 
Core i7 4770k procesor with 16GB RAM, while the second takes 75 minutes.
For convenience, these can be obtained online at 
[https://git.ecdf.ed.ac.uk/vallejosgroup/basicsworkflow2020](https://git.ecdf.ed.ac.uk/vallejosgroup/basicsworkflow2020).

```{r download-chain-naive}
if (!file.exists("rds/chain_naive.Rds")) {
  website <- "https://git.ecdf.ed.ac.uk/vallejosgroup/basicsworkflow2020/raw/master/"
  file <- "chain_naive.Rds"
  download.file(
    paste0(website, file),
    destfile = "rds/chain_naive.Rds"
  )  
}
if (!file.exists("rds/chain_active.Rds")) {
  website <- "https://git.ecdf.ed.ac.uk/vallejosgroup/basicsworkflow2020/raw/master/"
  file <- "chain_active.Rds"
  download.file(
    paste0(website, file),
    destfile = "rds/chain_active.Rds"
  )  
}
chain_naive <- readRDS("rds/chain_naive.Rds")
chain_active <- readRDS("rds/chain_active.Rds")
```

The output from `BASiCS_MCMC` is a `BASiCS_Chain` object that contains the 
draws associated to all model parameters (as `N` = 40,000, `Thin` = 20 and
`Burn` = 20,000, the object contains 1,000 draws for each parameter).
These can be accessed using the `displayChainBASiCS` function.
For example, the following code displays the first 2 draws for mean
expression parameters $\mu_i$ associated to the first 3 genes.

```{r displayChainBASiCS}
displayChainBASiCS(chain_naive, Param = "mu")[1:2, 1:3]
```

## MCMC diagnostics

Before interpreting the parameter estimates obtained by `r Biocpkg("BASiCS")`, 
it is critical to assess the convergence of the MCMC algorithm, i.e. whether
the MCMC reached its stationary distribution.
If convergence has been achieved, each parameter should not (on average) evolve
significantly across iterations, and MCMC draws are expected to be stochastic 
fluctuations around a horizontal trend.
Generally, it is not possible to prove convergence but multiple graphical and
quantitative convergence diagnostics have been proposed to assess the lack
of convergence (e.g. [@CowlesCarlin1996; @BrooksGelman1998]). 
Some advocate the use of multiple MCMC chains using different starting values in
order to ensure that the algorithm consistently converges to the same 
distribution.
For `r Biocpkg("BASiCS")`, we have observed that using informed starting values 
(e.g. based on `r Biocpkg("scran")` normalisation) and a sufficiently large 
value for `N` and `Burn` generally leads to consistent across multiple MCMC runs.
Hence, the focus of this section is to evaluate quantitative measures of 
convergence (e.g. [@Geweke1995]) based on a single MCMC chain.

Traceplots can be used to visually assess the history of MCMC iterations 
for a specific parameter (e.g. Figure \@ref(fig:convergence-naive)). 
As mentioned above, significan departures from a horizontal trend suggest a
lack of convergence.
As illustrated in Figure \@ref(fig:convergence-naive), histograms can also be 
used to display the marginal distribution for each parameter. 
Users should expect these to follow a unimodal distribution. 
Failure to satisfy these graphical diagnostics suggest that `N` and `Burn` must 
be increased. 
Alternatively, more stringent quality control could be applied to the input 
data, as genes with very low counts often suffer from slow convergence and poor sampling efficiency.

```{r convergence-naive, fig.cap="Trace plot, marginal histogram, and autocorrelation function for a gene in naive cells following MCMC sampling. Trace plots should explore the posterior well, without getting stuck in one location or drifting over time towards a region of higher density. High autocorrelation indicates that the number of effective independent samples is low. It is good practice to perform these visualisation for many different parameters; here we only show one."}
plot(chain_naive, Param = "mu", Gene = 1)
```

As `r Biocpkg("BASiCS")` infers thousands of parameters, it is 
impractical to assess these diagnostics for each parameter. 
Thus, it is helpful to use numerical diagnostics, applied to multiple
parameters simultaneously.
Here, we focus on the diagnostic criterion proposed by Geweke [@Geweke1995]
which compares the average draws obtained during the initial (10\% after burn 
in, by default) and the final part of the chain (50\% by default).
Large absolute $Z$-scores suggest that the algorithm has not converged. 
For the naive and activatived CD4^+^ T datasets most Z-scores associated to mean
expression parameters $\mu_i$ were small in aboslute value (see 
Figure \@ref(fig:geweke-diag)).

<!--- CAV: I removed the next sentence, trend is not that clear
after applying the additional gene filter. 

Further, as seen in Figures \@ref(fig:geweke-diag-mean) and
\@ref(fig:geweke-diag-detected), genes with large absolute values for this 
criterion are often those genes that have a low average expression level, or 
that are expressed only in a small proportion of cells.
---> 

```{r geweke-diag, fig.cap="Geweke Z-score for mean expression parameters is plotted against mean expression estimates for naive (left) and activated (right) cells.", eval = FALSE}
library("coda")
library("ggplot2")
library("viridis")

# Calculate and plot Geweke Z scores 
gew_mu_naive <- geweke.diag(mcmc(displayChainBASiCS(chain_naive, Param = "mu")))$z
gew_mu_active <- geweke.diag(mcmc(displayChainBASiCS(chain_active, Param = "mu")))$z

myplot_params <- list(
  geom_hex(),
  geom_hline(yintercept = c(-3, 3), col = "firebrick", linetype = "dashed"),
  scale_fill_viridis(),
  labs(x = "log(mu)", y = "Geweke Z-score"))

p_geweke_naive <- ggplot() +
  aes(
    log10(colMedians(displayChainBASiCS(chain_naive, Param = "mu"))),
    gew_mu_naive
  ) +
  myplot_params +
  ggtitle("Naive cells")
p_geweke_active <- ggplot() +
  aes(
    log10(colMedians(displayChainBASiCS(chain_active, Param = "mu"))),
    gew_mu_active
  ) +
  myplot_params+
  ggtitle("Activated cells")

p_geweke_naive + p_geweke_active
```

As well as assessing MCMC convergence, it is important to ensure that the MCMC 
algorithm has efficiently explored the parameter space. 
For example, the autocorrelation function (e.g. Figure 
\@ref(fig:convergence-naive), right panel) can be used to quantify the 
correlation between the chain and its lagged versions. 
Strong autocorrelation indicates that neighbouring MCMC samples are highly 
dependent and suggest poor sampling efficiency. 
The latter may indicate that the MCMC draws do not contain 
sufficient information to produce accurate posterior estimates.
In other words, highly correlated MCMC samplers require more samples to produce
the same level of Monte Carlo error for an estimate 
(defined as the variance of a Monte Carlo estimate across repetitions [@Koehler2009]). 

The effective sample size (ESS) is a related measure which represents a proxy 
for the number of independent draws generated by the MCMC sampler [@Gelman2014].
The latter is defined as:
$$
  \mbox{ESS} = \frac{N_{tot}}{1 + 2\sum_{k=1}^\infty \rho(k)},
$$
where $N_{tot}$ represents the total number of MCMC draws (after burn-in and
thining) and $\rho(k)$ $\rho(k)$ is the autocorrelation at lag $k$.
ESS estimates associated to mean expression parameters for the naive and 
activated CD4^+^ T cells are displayed in Figure \@ref(fig:ess-diag).
Whilst ESS is around 1,000 ($N_{tot}$ in this case) for most genes, we observe 
low ESS values for a small proportion of genes (primarily lowly expressed genes 
whose expression was only captured in a small number of cells).
As described later in this manuscript, `BASiCS_TestDE` automatically excludes 
genes with low ESS when performing differential expression testing.

```{r ess-diag, fig.cap="ESS for mean expression parameters is plotted against mean expression estimates for naive (left) and activated (right) cells.", eval = FALSE}
ess_mu_naive <- BASiCS_DiagPlot(chain_naive, Param = "mu") +
  theme(legend.position = "bottom")
ess_mu_active <- BASiCS_DiagPlot(chain_active, Param = "mu") +
  theme(legend.position = "bottom")
ess_mu_naive + ess_mu_active
```


## Quantifying transcriptional variability using BASiCS

Studying gene-level transcriptional variability can provide insights about the 
regulation of gene expression, and how it relates to the properties of genomic 
features (e.g. CpG island composition [@Morgan2018]), transcriptional dynamics 
[@Antolovic2017] and aging [@Martinez-jimenez2017], among others. 
The squared coefficient of variation (CV^2^) is widely used as a proxy for 
transcriptional variability. 
For example, here we obtain CV^2^ estimates for each gene using 
`r Biocpkg("scran")` normalised expression counts as input. 
Instead, `r Biocpkg("BASiCS")` infers transcriptional variability using 
gene-specific over-dispersion parameters $\delta_i$ (see *Methods*). 
Here, we compare these approaches focusing on naive CD4^+^ T cells (the 
analysis for active cells shows similar results). 
As seen in Figure \@ref(fig:dispersion-vs-mean)A, CV^2^ and posterior estimates 
for $\delta_i$ are highly correlated. 
Moreover, both variability metrics are confounded by differences in mean 
expression, i.e. highly expressed genes tend to exhibit lower variability
(Figures \@ref(fig:dispersion-vs-mean)B-C). 
To remove this confounding, `r Biocpkg("scran")` and `r Biocpkg("BASiCS")`
derive *residual variability* estimates as deviations with respect to an
global mean-variability trend (see *Methods*).
These are derived using the DM approach [@Kolodziejczyk2015cell] and the
residual over-dispersion parameters $\epsilon_i$ defined by [@Eling2018],
respectively.
For the naive CD4^+^ T data, both approaches led to strongly correlated 
estimates (Figure \@ref(fig:dispersion-vs-mean)D) and, as expected, neither DM 
values or posterior estimates for $\epsilon_i$ are seen to be associated with
mean expression (Figure \@ref(fig:dispersion-vs-mean)E-F). 
However, unlike the DM method, the integrated approach implemented in 
`r Biocpkg("BASiCS")` provides a direct measure of statistical uncertainty for 
these estimates. 
As illustrated in the following sections, this enables gene-specific
downstream inference such as differential variability analyses. 

<!--- CV: Can we make the n_neigh smaller? --->

 <!--- CV: add this when editing caption: Note that the `BASiCS_ShowFit` function can be used to generate panel B, but we generated the plot manually to demonstrate how users can extract this information from a `BASiCS_MCMC` object. These genes have missing (`NA`) values for the residual over-dispersion parameter, $\epsilon$. We use `r CRANpkg("ggpointdensity")` to visualise the local density of genes along the axes of mean and variability. ---> 
 
```{r dispersion-vs-mean, fig.width=6, fig.height=6, fig.cap="Comparison of gene-specific transcriptional variability estimates and mean expression estimates obtained for each gene using BASiCS and scran. For this analysis, we exclude genes that are not expressed in at least 2 cells. BASiCS estimates for each gene are defined by the posterior median of the associated parameter. scran estimates for each gene are derived after applying the pooling normalisation strategy proposed by Lun et al. Points are coloured according to the local density of genes along the x- and y- axis. A: scran squared CV estimates versus BASiCS estimates for over-dispersion parameters. B: scran estimates for mean expression and the squared CV. C: BASiCS estimates for mean expression and over-dispersion parameters. D: BASiCS estimates for residual over-dispersion parameters versus DM values estimated by scran. E: scran estimates for mean expression and DM values. F: BASiCS estimates for mean expression and residual over-dispersion parameters."}

library("ggpointdensity")
library("viridis")

# Get BASiCS posterior estimates for mean and variability - naive cells
summary_naive <- Summary(chain_naive)
parameter_df <- data.frame(
  mu = displaySummaryBASiCS(summary_naive, Param = "mu")[, 1],
  delta = displaySummaryBASiCS(summary_naive, Param = "delta")[, 1],
  epsilon = displaySummaryBASiCS(summary_naive, Param = "epsilon")[, 1]
)

# Get scran estimates for mean and variability - naive cells
sce_naive <- logNormCounts(sce_naive, log = FALSE)
parameter_df$mean_scran <- rowMeans(assay(sce_naive, "normcounts"))
parameter_df$cv2_scran <- rowVars(assay(sce_naive, "normcounts")) /
  parameter_df$mean_scran^2
parameter_df$DM <- DM(
  mean = parameter_df$mean_scran, 
  cv2 = parameter_df$cv2_scran
)

# Remove genes without counts in > 2 cells - BASiCS estimates not provided
ind_not_na <- !(is.na(parameter_df$epsilon))

myplot_params <- list(geom_pointdensity(),
                      scale_colour_viridis(),
                      theme(
                        text = element_text(size=rel(3)),
                        legend.position = "bottom",
                        legend.text = element_text(angle = 45, size = 8),
                        legend.key.size =  unit(0.8, "cm")))

g1 <- ggplot(parameter_df[ind_not_na, ], aes(log10(cv2_scran), log10(delta))) +
  myplot_params +
  xlab("scran CV^2 (log10)") +
  ylab("BASiCS\nover-dispersion (log10)") +
  geom_abline(slope = 1, intercept = 0, col = "red")

g2 <- ggplot(parameter_df[ind_not_na, ], aes(log10(mean_scran), log10(cv2_scran))) +
  myplot_params +
  xlab("scran means (log10)") +
  ylab("scran CV^2 (log10)")

g3 <- ggplot(parameter_df[ind_not_na, ], aes(log10(mu), log10(delta))) +
  myplot_params +
  xlab("BASiCS means (log10)") +
  ylab("BASiCS\nover-dispersion (log10)")

g4 <- ggplot(parameter_df[ind_not_na, ], aes(DM, epsilon)) +
  myplot_params +
  xlab("scran DM") +
  ylab("BASiCS residual\nover-dispersion") +
  geom_abline(slope = 1, intercept = 0, col = "red")

g5 <- ggplot(parameter_df[ind_not_na, ], aes(log10(mean_scran), DM)) +
  myplot_params +
  xlab("scran means (log10)") +
  ylab("scran DM") +
  geom_hline(yintercept = 0, col = "red")

g6 <- ggplot(parameter_df[ind_not_na, ], aes(log10(mu), epsilon)) +
  myplot_params +
  xlab("BASiCS means (log10)") +
  ylab("BASiCS residual\nover-dispersion") +
  geom_hline(yintercept = 0, col = "red")

(g1 + g2 + g3) / (g4 + g5 + g6) + 
  plot_annotation(tag_levels = "A") & theme(plot.tag = element_text(size = 15))
```

## HVG/LVG detection using BASiCS

In `r Biocpkg("BASiCS")`, the functions `BASiCS_DetectHVG` and 
`BASiCS_DetectLVG` can be used to identify genes with substantially high (HVG) 
or low (LVG) transcriptional variability within a population of cells. 
If the input `BASiCS_Chain` object was generated by `BASiCS_MCMC` with 
`Regression = TRUE` (recommended setting), this analysis is based on the 
posterior distribution obtained for gene-specific residual over-dispersion 
parameters $\epsilon_i$ (alternatively, the approach introduced by 
[@Vallejos2015] can be used). 
HVGs are marked as those for which $\epsilon_i$ exceeds a pre-defined threshold 
with high probability, where the probability cut-off is chosen to match a given 
expected false discovery rate (EFDR; default EFDR = 10\%) [@Newton2004].
A similar approach is implemented for LVG detection, but based on whether 
$\epsilon_i$ is below a pre-specified threshold. 
In principle, a threshold for $\epsilon_i$ could be directly specified by the 
user. 
For example, if the threshold for $\epsilon_i$ is equal to $\log(2)$, HVGs would 
be those genes for which the over-dispersion is estimated to be at least $2$ 
times higher than would be expected given the inferred mean expression level. 
In practice, however, it may be of interest to rank genes and to select the
those with the highest or the lowest residual over-dispersion. 
<!--- CV: LIKELY TO BE EDITED TO USE THE APPROACH ABOVE --->
For this purpose, the `PercentileThreshold` parameter can be used to define the 
threshold based on the empirical distribution of posterior estimates for 
residual over-dispersion parameters $\epsilon_i$. 
For example, if `PercentileThreshold = 0.9`, the `BASiCS_DetectHVG` function 
defines the threshold as the 90\% percentile of the above distribution.


```{r plot-vg-naive, fig.cap="HVG and LVG detection using BASiCS. For each gene, BASiCS posterior estimates (posterior medians) associated to mean expression and residual over-dispersion parameters are plotted. Genes are coloured according to HVG/LVG status. Genes that are not expressed in at least 2 cells are excluded."}
# Highly variable genes
HVG <- BASiCS_DetectHVG(chain_naive, PercentileThreshold = 0.9)
# Lowly variable genes
LVG <- BASiCS_DetectLVG(chain_naive, PercentileThreshold = 0.1)

ggplot(HVG@Table) +
  geom_point(aes(log10(Mu), Epsilon), na.rm = TRUE, colour = "gray") +
  xlab("BASiCS means (log10)") +
  ylab("BASiCS residual\nover-dispersion") +
  geom_hline(yintercept = 0, lty = 2) +
  geom_point(data = HVG@Table[HVG@Table$HVG == TRUE, ],
             aes(log10(Mu), Epsilon), na.rm = TRUE, colour = "red") +
  geom_point(data = LVG@Table[LVG@Table$LVG == TRUE, ],
             aes(log10(Mu), Epsilon), na.rm = TRUE, colour = "blue")
```

For the naive CD4^+^ T data, we obtained *`r sum(HVG@Table$HVG, na.rm=TRUE)`* 
HVG and *`r sum(LVG@Table$LVG, na.rm=TRUE)`* LVG.
As shown in Figure \@ref(fig:plot-vg-naive), these genes are distributed across 
a wide range of mean expression values. 
As an illustration, Figure \@ref(fig:plot-example-vg-naive) shows the 
distribution of normalised expression values for selected HVG and LVG, focusing 
on examples with similar mean expression levels (Figure 
\@ref(fig:plot-example-vg-naive)A). 
As expected HVG tend to exhibit a wider distribution and potentially bimodal
distribution (Figure \@ref(fig:plot-example-vg-naive)B).
Instead, LVG tend to have more narrow and unimodal distributions 
(Figure \@ref(fig:plot-example-vg-naive)C).


```{r plot-example-vg-naive, fig.cap = "BASiCS denoised counts for example HVG and LVG with similar overall levels of expression."}

library(reshape2)

# Obtain normalised expression values
DC_naive <- BASiCS_DenoisedCounts(sce_naive, chain_naive)

HVG_Table <- merge(HVG@Table, genenames, 
                   by.x = "GeneName", by.y = "ensembl_gene_id", 
                   all.x = TRUE, sort = FALSE)
LVG_Table <- merge(LVG@Table, genenames, 
                   by.x = "GeneName", by.y = "ensembl_gene_id", 
                   all.x = TRUE, sort = FALSE)

# Select HVG/LVG genes with similar mean expression values  
low.exp <- 2.8
up.exp <- 3
HVG1 <- HVG_Table[HVG_Table$HVG == TRUE &
                    log10(HVG_Table$Mu) > low.exp & 
                    log10(HVG_Table$Mu) < up.exp,]
LVG1 <- LVG_Table[LVG_Table$LVG == TRUE &
                    log10(LVG_Table$Mu) > low.exp & 
                    log10(LVG_Table$Mu) < up.exp,]  

# Order by epsilon and select top 3 HVG and LVG within the genes selected above
HVG1 <- HVG1[order(HVG1$Epsilon, decreasing = TRUE),]
LVG1 <- LVG1[order(LVG1$Epsilon, decreasing = FALSE),]
topHVG <- log10(t(DC_naive[HVG1$GeneName[1:3], ]) + 1)
topLVG <- log10(t(DC_naive[LVG1$GeneName[1:3], ]) + 1)
# Add genenames
colnames(topHVG) <- HVG1$external_gene_name[1:3]
colnames(topLVG) <- LVG1$external_gene_name[1:3]
  
myplot_params <- list(geom_violin(na.rm = TRUE),
                      coord_flip(),
                      ylim(-0.05, max(log10(DC_naive + 1))),
                      geom_jitter(position=position_jitter(0.3)),
                      ylab("log10(norm count + 1)"),
                      xlab("Gene"))
plot_hvg <- ggplot(melt(topHVG), aes(x = Var2, y = value)) + 
  myplot_params + ggtitle("Example HVG")

plot_lvg <- ggplot(melt(topLVG), aes(x = Var2, y = value)) + 
  myplot_params + ggtitle("Example LVG")
  
plot_hvg + plot_lvg + plot_annotation(tag_levels = "A") 
```

## Differential mean and variability testing using BASiCS

This section highlights the use of `r Biocpkg("BASiCS")` to perform differential 
expression tests for mean and variability between different pre-specified 
populations of cells and experimental conditions. 
Here, we compare the naive CD4^+^ T cells, analysed in the previous section, to 
activated CD4^+^ T cells obtained by the same authors [@Martinez-jimenez2017]. 
Naive CD4^+^ T cells were activated for 3 hours 
using plate-bound CD3e and CD28 antibodies.
T cell activation is linked to strong transcriptional shifts and the
up-regulation of lineage specific marker genes, such as Tbx21 and Gata1 
[@Best2013;@Fu2012].
To generate this data, the authors did not add cytokines, which are needed for 
T cell differentiation [@Zhu2010]. Therefore, any heterogeneity in the activated
cell population does not arise from cells residing in different lineage-specific 
differentiation states. 

`r Biocpkg("BASiCS")` implements differential expression testing via the
`BASiCS_TestDE` function. The main input parameters for `BASiCS_TestDE` are:

<!-- Removed testing over-dispersion from here, not the main feature (e.g. naive and active CD4^+^ T cells). ---> 
<!--`EpsilonD` ---> 
- `Chain1` and `Chain2`: two `BASiCS_Chain` objects created via the `BASiCS_MCMC`
function. Each object corresponds to a different pre-specified group of cells
- `EpsilonM` and `EpsilonR`: introduce a minimum effect size (in a
$\log_2$ fold change scale) for the detection of changes in mean or residual 
over-dispersion, respectively. 
This enables us to discard small expression changes that are less biologically
meaningful.
By default, we set these thresholds to be equivalent to a 50\% change between
the groups. 
However, different thresholds may be required depending on the context.
For example, if most genes show strong differences in mean expression, it can be 
beneficial to increase the value of `EpsilonM` to focus on strong changes in 
mean expression. 
- `EFDR_M` and `EFDR_R`: define the target EFDR to calibrate the 
decision rule associated to changes in mean or residual over-dispersion, 
respectively. Default is 10\%.  
- `MinESS`: Minimum ESS threshold beyond which genes will be excluded from the
differential expression tests. 
This is used to increase the robustness of the results, excludes genes for which 
the sampler explored the parameter space less efficiently (see 
*MCMC diagnostics* Section). Default: `MinESS` = 100. 

```{r mean-expression-testing}
# Perform differential testing
Test_DE <- BASiCS_TestDE(
  Chain1 = chain_naive,
  Chain2 = chain_active,
  GroupLabel1 = "Naive",
  GroupLabel2 = "Active",
  Plot = FALSE,
  PlotOffset = FALSE
)
TableMean <- format(Test_DE, Which = "Mean", Filter = FALSE)
TableResDisp <- format(Test_DE, Which = "ResDisp", Filter = FALSE)
```

After running the test, it is important to visualise the results in order to
interpret the results and identify systematic patterns among differentially
expressed genes. 
It may also be useful to perform functional enrichment analysis to identify 
biologically meaningful patterns amongst differentially expressed genes. 
For example, this could be done using `r Biocpkg("goseq")` [@Young2010].
We do not perform this here, as it is outside of the scope of 
`r Biocpkg("BASiCS")`. 
However, a workflow for this process is detailed in Supplementary section TODO. 
<!--- INSTEAD OF SUPP, IS THERE ANOTHER WORKFLOW TO CITE HERE? ---> 

Here we first focus on the differential mean expression test. 
MA-plots (log fold change M versus mean average A) and volcano plots (posterior 
probability versus log fold change) are popular graphical summaries in this
context (see Figure \@ref(visualise-DE-mean-plot)). 

```{r visualise-DE-mean-plot, fig.height = 8, fig.width = 6, fig.cap = "Upper panel presents the MA plot associated to the differential mean expression test between naive and active cells. Log-fold changes of average expression in naive cells relative to active cells are plotted against average expression estimates combined across both groups of cells. Bottom panel presents the volcano plot associated to the same test. Log-fold changes of average expression in naive cells relative to active cells are plotted against their associated tail posterior probabilities. Colour indicates the differential expression status for each gene, including a label to identify genes that were excluded from differential expression test due to low ESS."}
p1 <- BASiCS_PlotDE(Test_DE, Parameters = "Mean", Plots = "MA") 
p2 <- BASiCS_PlotDE(Test_DE, Parameters = "Mean", Plots = "Volcano") 
p1 / p2
```

When interpreting the results of differential expression tests, it is useful
to visualise expression patterns for differentially expressed genes in order to 
appraise the significance of the results and guide interpretation. 
For this purpose, we obtain normalised expression values for each group of 
cells after correcting for global changes in overall expression between the 
groups, i.e. a global offset that leads to uniformly higher expression across 
genes within one group of cells (such step is also internally done by 
`BASiCS_TestDE`). 
Among others, the latter can be due to overall differences in mRNA content
between the groups. 

```{r offset-denoised}
# Calculate a global offset between the groups
my_offset <- BASiCS_CorrectOffset(chain_naive, chain_active)
my_offset$Offset
# Get offset corrected chain
chain_naive <- my_offset$Chain
# Obtain normalised counts within each group of cells
DC_naive <- BASiCS_DenoisedCounts(sce_naive, chain_naive)
DC_active <- BASiCS_DenoisedCounts(sce_active, chain_active)
```

To visualise expression patterns for multiple genes at once, we use the 
`r Biocpkg("ComplexHeatmap")` [@Gu2016] Biocondutor package, grouping genes
according to the result of the differential mean expression test (i.e. 
up-regulated in naive/active cells or non differentially expressed). 
Figure \@ref(fig:heatmap-diffexp) shows 15 genes are shown for each group. 
Fpr example, amongst the non DE group, we observe several genes ribosomal 
proteins (e.g. *Rps14*). 
Genes in this family have been previously observed to have stable expression
across a wide range of scRNAseq datasets in mouse and human [REF]. 
Moreover, the genes that are found to be upregulated in the active group include 
known markers of T cell activation such as **Cd69* [@Ziegler1994]. 


```{r heatmap-diffexp, fig.width = 7, fig.height = 6, fig.cap="Heatmap displays normalised expression values (log10(x + 1) scale) for naive and active cells. Genes are stratified according to their differential expression status (non differentially expressed; upregulated in naive or active cells). For each group, 15 example genes are shown. These were selected according to the ranking of their associated tail posterior probabilities associated to the differential mean expression test. Colour indicates expression level; colour bars on the right of heatmap segments indicate the inferred mean expression level (in log scale) for each gene in each population."}
library("ComplexHeatmap")
library("circlize")
library("RColorBrewer")

# Up-regulated in naive
ind_n <- TableMean$ResultDiffMean == "Naive+"
naive_mean <- TableMean[ind_n, ]
naive_mean <- naive_mean[order(naive_mean$MeanLog2FC, decreasing = FALSE), ]
naive_mean$Symbol <- genenames[naive_mean$GeneName, 2]

# Up-regulated in active
ind_a <- TableMean$ResultDiffMean == "Active+"
active_mean <- TableMean[ind_a, ]
active_mean <- active_mean[order(active_mean$MeanLog2FC, decreasing = TRUE), ]
active_mean$Symbol <- genenames[active_mean$GeneName, 2]

# Non DE genes
ind_non <- TableMean$ResultDiffMean == "NoDiff"
nonde_mean <- TableMean[ind_non, ]
nonde_mean <- nonde_mean[order(abs(nonde_mean$MeanLog2FC), decreasing = FALSE), ]
nonde_mean$Symbol <- genenames[nonde_mean$GeneName, 2]

heatmap_ngenes <- 15
heatmap_seq <- seq_len(heatmap_ngenes)

## Select genes with largest probability of differential expression
active_ind <- order(active_mean$ProbDiffMean, decreasing = TRUE)[heatmap_seq]
active_mean <- active_mean[active_ind, ]
## subset counts from each cell type to these genes
act_counts_act <- DC_active[active_mean$GeneName, ]
nai_counts_act <- DC_naive[active_mean$GeneName, ]

## Select genes with largest probability of differential expression
s_ind <- order(naive_mean$ProbDiffMean, decreasing = TRUE)[heatmap_seq]
naive_mean <- naive_mean[s_ind, ]
## subset counts from each cell type to these genes
act_counts_nai <- DC_active[naive_mean$GeneName, ]
nai_counts_nai <- DC_naive[naive_mean$GeneName, ]

## Select genes with lowest probability of differential expression
s_ind <- order(nonde_mean$ProbDiffMean, decreasing = FALSE)[heatmap_seq]
nonde_mean <- nonde_mean[s_ind, ]
## subset counts from each cell type to these genes
act_counts_nde <- DC_active[nonde_mean$GeneName, ]
nai_counts_nde <- DC_naive[nonde_mean$GeneName, ]

## Calculate max absolute value for min/max of colour scale
all_mean <- c(
  active_mean$Mean1, active_mean$Mean2,
  naive_mean$Mean1, naive_mean$Mean2,
  nonde_mean$Mean1, nonde_mean$Mean2
)
max_range <- log(max(abs(all_mean)))
## colour scale symmetric around zero
colour <- colorRamp2(seq(-max_range, max_range, length.out = 9),
                     rev(brewer.pal(9, "RdBu")))

## Combine count matrices by cell type
counts_active <- rbind(act_counts_act, act_counts_nai, act_counts_nde)
counts_naive <- rbind(nai_counts_act, nai_counts_nai, nai_counts_nde)

## split heatmaps by gene category
split <- data.frame(
  Upregulated = c(
    rep("Up-regulated \nin active", nrow(act_counts_act)),
    rep("Up-regulated \nin naive", nrow(act_counts_nai)),
    rep("Non DE", nrow(act_counts_nde))
  )
)
syms <- genenames[rownames(counts_active), 2]
fontsize <- 7

Heatmap(
  log10(counts_naive + 1),
  row_labels = syms,
  row_names_gp = gpar(fontsize = fontsize),
  name = "log10(count + 1)",
  column_dend_height = unit(0.2, "npc"),
  column_title_side = "bottom",
  column_title = "Naive cells",
  show_column_names = FALSE,
  cluster_rows = FALSE,
  split = split,
  right_annotation = rowAnnotation(
    log_mu = log(c(active_mean$Mean1, naive_mean$Mean1, nonde_mean$Mean1)),
    col = list(log_mu = colour)
  ),
  col = viridis(100)) +
  Heatmap(
    log10(counts_active + 1),
    row_labels = syms,
    column_dend_height = unit(0.2, "npc"),
    row_names_gp = gpar(fontsize = fontsize),
    column_title = "Active cells",
    column_title_side = "bottom",
    show_column_names = FALSE,
    #name = "log10(count + 1)",
    split = split,
    right_annotation = rowAnnotation(
      log_mu = log(c(active_mean$Mean2, naive_mean$Mean2, nonde_mean$Mean2)),
      col = list(log_mu = colour)
    ),
    cluster_rows = FALSE,
    col = viridis(100))
```

Whist several computational tools exist to perform differential mean expression
analysis using scRNAseq data [@Soneson2018], they key focus of
`r Biocpkg("BASiCS")` is to perform differential variability testing:
identifying changes in transcriptional variability between the groups of cells. 
To avoid the mean versus over-dispersion confounding, we recommend this analysis
to use residual over-dispersion parameters $\epsilon_i$ as input. 

We can now visualise the changes in residual over-dispersion between naive and 
activated CD4^+^ T cells in the form of a MA-plot 
(Figure \@ref(fig:diff-res-plot)).
In this visualisation, the difference between the posterior medians of the 
residual over-dispersion parameters $\epsilon$ are shown on the y-axis.
Epsilon values for genes that are not expressed in at least 2 cells per 
conditions are marked as `NA` and are therefore not being displayed.


```{r diff-res-plot, fig.height = 8, fig.width = 6, fig.cap = "Upper panel presents the MA plot associated to the differential residual overdispersion test between naive and active cells. Differences of residual overdispersion in naive cells relative to active cells are plotted against average expression estimates combined across both groups of cells. Bottom panel presents the volcano plot associated to the same test. Differences of residual overdispersion in naive cells relative to active cells are plotted against their associated tail posterior probabilities. Colour indicates the differential expression status for each gene, including a label to identify genes that were excluded from differential expression test due to low ESS."}
TableResDisp <- format(Test_DE, Which = "ResDisp", Filter = FALSE)
p1 <- BASiCS_PlotDE(Test_DE, Parameters = "ResDisp", Plots = "MA") 
p2 <- BASiCS_PlotDE(Test_DE, Parameters = "ResDisp", Plots = "Volcano") 
p1 / p2
```

While one could focus on the genesets that show significant changes in
residual over-dispersion, here, we want to highlight how to analyse 
changes in mean expression in parallel to changes in variability.
For this, we will first combine the results of the differential mean expression 
and the differential residual over-dispersion test.
<!---We will further remove the genes that were excluded from the test and those 
that are not expressed in at least 2 cells in either condition.---> 

```{r combine-results}
TableCombined <- merge(TableMean, TableResDisp, by = "GeneName")
```

As discussed [previously](#vis-diff-exp), it is useful to visualise the results
of differential expression tests in order to appraise the quality of the 
results, and to identify systematic patterns among the genes identified.
For this purpose, we again use `r Biocpkg("ComplexHeatmap")`; the result is
shown in Figure \@ref(fig:heatmap-diffresdisp).

```{r heatmap-diffresdisp, fig.width = 7, fig.height = 5, fig.cap="Heatmap of 30 genes with higher residual over-dispersion in naive or active cells. Colour indicates expression level; colour bars on the right of heatmap segments indicate the residual over-dispersion for each gene in that population."}
# More variable in naive, but non DE
ind_n <- TableCombined$ResultDiffResDisp == "Naive+" &
  TableCombined$ResultDiffMean == "NoDiff"
naive_resdisp <- TableCombined[ind_n, ]
ord <- order(naive_resdisp$ResDispDistance, decreasing = FALSE)
naive_resdisp <- naive_resdisp[ord, ]
naive_resdisp$Symbol <- genenames[naive_resdisp$GeneName, 2]

# More variable in active, but non DE
ind_a <- TableCombined$ResultDiffResDisp == "Active+" & 
  TableCombined$ResultDiffMean == "NoDiff"
active_resdisp <- TableCombined[ind_a, ]
ord <- order(active_resdisp$ResDispDistance, decreasing = TRUE)
active_resdisp <- active_resdisp[ord, ]
active_resdisp$Symbol <- genenames[active_resdisp$GeneName, 2]

heatmap_ngenes <- 15
heatmap_seq <- seq_len(heatmap_ngenes)

## Select genes with largest probability of differential variability
active_ind <- order(active_resdisp$ProbDiffResDisp, decreasing = TRUE)[heatmap_seq]
active_resdisp <- active_resdisp[active_ind, ]
## Subset count matrices to genes with highest probability
act_counts_act <- DC_active[active_resdisp$GeneName, ]
nai_counts_act <- DC_naive[active_resdisp$GeneName, ]

## Select genes with largest probability of differential variability
s_ind <- order(naive_resdisp$ProbDiffResDisp, decreasing = TRUE)[heatmap_seq]
naive_resdisp <- naive_resdisp[s_ind, ]
## Subset count matrices to genes with highest probability
act_counts_nai <- DC_active[naive_resdisp$GeneName, ]
nai_counts_nai <- DC_naive[naive_resdisp$GeneName, ]

## Calculate max values for colour scale
all_resdisp <- c(
  active_resdisp$ResDisp1, active_resdisp$ResDisp2,
  naive_resdisp$ResDisp1, naive_resdisp$ResDisp2
)
max_range <- max(abs(all_resdisp))
## colour scale symmetric around zero
colour <- colorRamp2(seq(-max_range, max_range, length.out = 9),
  rev(brewer.pal(9, "RdBu")))

## Combine count matrices by cell type
counts_active <- rbind(act_counts_act, act_counts_nai)
counts_naive <- rbind(nai_counts_act, nai_counts_nai)

## Split heatmaps using gene categories
split <- data.frame(
  Upregulated = c(
    rep("More variable \nin active", nrow(act_counts_act)),
    rep("More variable \nin naive", nrow(act_counts_nai))
  )
)
syms <- genenames[rownames(counts_active), 2]

fontsize <- 7

Heatmap(
  log10(counts_naive + 1),
  row_labels = syms,
  row_names_gp = gpar(fontsize = fontsize),
  #name = "log10(count + 1)",
  column_dend_height = unit(0.2, "npc"),
  column_title_side = "bottom",
  column_title = "Naive cells",
  show_column_names = FALSE,
  cluster_rows = FALSE,
  split = split,
  right_annotation = rowAnnotation(
    epsilon = c(active_resdisp$ResDisp1, naive_resdisp$ResDisp1),
    col = list(epsilon = colour)
  ),
  col = viridis(100)) +
  Heatmap(
    log10(counts_active + 1),
    row_labels = syms,
    column_dend_height = unit(0.2, "npc"),
    row_names_gp = gpar(fontsize = fontsize),
    column_title = "Active cells",
    column_title_side = "bottom",
    show_column_names = FALSE,
    name = "log10(count + 1)",
    split = split,
    right_annotation = rowAnnotation(
      epsilon = c(active_resdisp$ResDisp2, naive_resdisp$ResDisp2),
      col = list(epsilon = colour)
    ),
    cluster_rows = FALSE,
    col = viridis(100))
```

<!--- NOTE: heatmap doesn't seem to be the best way to visualise results.
This section needs to be edited + some interpretation added ---> 

