---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Nils Eling
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
    email: eling@ebi.ac.uk
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
  - name: John C. Marioni
    affiliation: 
    - *EBI
    - *CRUK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems, such as immunity and development. Single-cell RNA 
  sequencing is a powerful tool to quantify this heterogeneity, but it is prone 
  to strong technical noise. In this article, we describe a step-by-step 
  computational workflow which uses the BASiCS Bioconductor package to robustly 
  quantify expression variability within and between known groups of cells (such 
  as experimental conditions or cell types). BASiCS uses an integrated framework 
  for data normalisation, technical noise quantification and downstream 
  analyses, whilst propagating statistical uncertainty across these steps. 
  Within a single seemingly homogeneous cell population, BASiCS can identify 
  highly variable genes that exhibit strong heterogeneity as well as lowly 
  variable genes with stable expression. BASiCS also uses a probabilistic 
  decision rule to identify changes in expression variability between cell 
  populations, whilst avoiding confounding effects related to differences in 
  technical noise or in overall abundance. Using two publicly available 
  datasets, we guide users through a complete pipeline which includes 
  preliminary steps for quality control as well as data exploration 
  using the scater and scran Bioconductor packages. Data for the first case 
  study was generated using the Fluidigm\@ C1 system, in which extrinsic 
  spike-in RNA molecules were added as a control. The second dataset was 
  generated using a droplet-based system, for which spike-in RNA is not 
  available. This analysis provides an example, in which differential 
  variability testing reveals insights regarding a possible early cell fate 
  commitment process. The workflow is accompanied by a Docker image that 
  ensures the reproducibility of our results. 
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article: default
  BiocStyle::html_document: default
---


```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
# Use fig.width = 7 for html and fig.width = 6 for pdf
fig.width <- ifelse(on.bioc, 7, 6)
knitr::opts_chunk$set(
  cache = TRUE, warning = FALSE, message = FALSE, error = FALSE,
  cache.path = "cache/", fig.path = "figure/", fig.width = fig.width
)
```

## Reproducibility

The following R, Bioconductor and package version were used for this workflow:

**R version**: `r R.version.string`

**Bioconductor version**: `r BiocManager::version()`

**Packages**: BASiCS `r packageVersion("BASiCS")`, scran 
`r packageVersion("scran")`, scater `r packageVersion("scater")`, coda 
`r packageVersion("coda")`, goseq `r packageVersion("goseq")`    

For the full list of packages used, please see the [Session Info](#session-info).

# C1 Fluidigm data: Analysis of naive CD4^+^ T cells {#Tcells}

For the first case study, we will use scRNA-seq data of CD4^+^ T cells, which 
were processed using the C1 Single-Cell Auto Prep System (Fluidigm^®^) using 
10–17 $\mu{}m$ integrated fluidic circuits (IFCs). 
Martinez-Jimenez _et al._ profiled naive and activated CD4^+^ T cells from 
young and old animals across two mouse strains to test for changes in 
expression variability that occur during organismal ageing 
[@Martinez-jimenez2017]. They extracted naive or effector memory CD4^+^ T cells
from spleens of young or old animals and filtered using either 
magnetic-activated cell sorting (MACS) or fluorescence activated cell sorting 
(FACS) (labelled as MACS-purified Naive, FACS-purified Naive or FACS-purified 
Effector Memory).
For clarification, naive CD4^+^ T cells are also referred to as 'unstimulated' 
CD4^+^ T cells.

In addition to profiling naive CD4^+^ T cells, the authors stimulated half 
of the naive cells for 3 hours using _in vitro_ antibody stimulation (labelled 
as Active).
They processed naive as well as activated CD4^+^ T cells using the 
C1 Fluidigm^®^ system to capture and lyse cells, and to reverse transcribe 
and amplify mRNA prior to sequencing.
The authors isolated cells from B6 (C57BL/6J, _Mus musculus domesticus_) and
CAST (_Mus musculus castaneus_) animals to profile the evolutionary conservation
of transcriptional dynamics during ageing.
Additionally, the authors added external spike-in RNA to aid in quantifying
technical variability across all cells. They performed all experiments in
replicates (also referred to as batches) to control for batch effects.

We will begin the workflow with obtaining the data before quality control, 
running the BASiCS model, and performing further downstream analysis.


## Obtaining the data

The raw counts of the full dataset can be obtained from ArrayExpress under the 
accession number 
[E-MTAB-4888](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-4888/).
In this dataset, the column names contain the library identifier of the original 
experiment, while the row names of the matrix store gene names. 
The dataset contains reads mapped to ERCC spike-in genes [@Rna2005], which 
`BASiCS` uses to estimate and remove technical noise.

