\documentclass[9pt,a4paper,]{extarticle}

\usepackage{f1000_styles}

\usepackage[pdfborder={0 0 0}]{hyperref}

\usepackage[numbers]{natbib}
\bibliographystyle{unsrtnat}


%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother


% disable code chunks background
%\renewenvironment{Shaded}{}{}

% disable section numbers
\setcounter{secnumdepth}{0}

%% added by MLS, this is not in the F1000 style by default %%

\hypersetup{unicode=true,
            pdftitle={BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data},
            pdfkeywords={Single-cell RNA sequencing, expression variability, transcriptional noise, differential expression testing},
            colorlinks=true,
            linkcolor=Maroon,
            citecolor=Blue,
            urlcolor=Orange,
            breaklinks=true}

%% End added by MLS %%

\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}



\begin{document}
\pagestyle{front}

\title{BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data}

\author[1,2]{Nils Eling\thanks{\ttfamily eling@ebi.ac.uk}}
\author[3]{Alan O'Callaghan}
\author[1,2]{John C. Marioni}
\author[3,4]{Catalina A. Vallejos\thanks{\ttfamily catalina.vallejos@igmm.ed.ac.uk}}
\affil[1]{European Molecular Biology Laboratory, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK}
\affil[2]{Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Cambridge, CB2 0RE, UK}
\affil[3]{MRC Human Genetics Unit, Institute of Genetics \& Molecular Medicine, University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, EH4 2XU, UK}
\affil[4]{The Alan Turing Institute, British Library, 96 Euston Road, London, NW1 2DB, UK}

\maketitle
\thispagestyle{front}

\begin{abstract}
Cell-to-cell gene expression variability is an inherent feature of complex
biological systems, such as immunity and development. Single-cell RNA
sequencing is a powerful tool to quantify this heterogeneity, but it is prone
to strong technical noise. In this article, we describe a step-by-step
computational workflow which uses the BASiCS Bioconductor package to robustly
quantify expression variability within and between known groups of cells (such
as experimental conditions or cell types). BASiCS uses an integrated framework
for data normalisation, technical noise quantification and downstream
analyses, whilst propagating statistical uncertainty across these steps.
Within a single seemingly homogeneous cell population, BASiCS can identify
highly variable genes that exhibit strong heterogeneity as well as lowly
variable genes with stable expression. BASiCS also uses a probabilistic
decision rule to identify changes in expression variability between cell
populations, whilst avoiding confounding effects related to differences in
technical noise or in overall abundance. Using two publicly available
datasets, we guide users through a complete pipeline which includes
preliminary steps for quality control as well as data exploration
using the scater and scran Bioconductor packages. Data for the first case
study was generated using the Fluidigm@ C1 system, in which extrinsic
spike-in RNA molecules were added as a control. The second dataset was
generated using a droplet-based system, for which spike-in RNA is not
available. This analysis provides an example, in which differential
variability testing reveals insights regarding a possible early cell fate
commitment process. The workflow is accompanied by a Docker image that
ensures the reproducibility of our results.
\end{abstract}

\section*{Keywords}
Single-cell RNA sequencing, expression variability, transcriptional noise, differential expression testing


\clearpage
\pagestyle{main}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

Single-cell RNA-sequencing (scRNA-seq) enables the study of genome-wide
transcriptional heterogeneity in cell populations that remains otherwise
undetected in bulk experiments \citep{Stegle2015, Prakadan2017, Patange2018}.
Applications of scRNA-seq range from characterising cell types in immunity
\citep{Lonnberg2017, Villani2017, Zheng2017} and development \citep{Ibarra-Soria2018, Wagner2018, Pijuan-Sala2019} to dissecting the mechanisms for cell fate
commitment \citep{Goolam2016, Ohnishi2014}.
Transcriptional heterogeneity within a population of cells can relate to
different underlying structures.
On the broadest level, this heterogeneity can relate to the presence of distinct
expression profiles associated to cell subtypes or discrete states, which
could be characterised through clustering \citep{Kiselev2019}.
Alternatively, cell-to-cell expression heterogeneity can reflect gradual
changes along processes that evolve over time and that can be characterised
using pseudotime inference methods \citep{Saelens2019}.
The focus of this article is on more subtle expression variability that can
occur within a seemingly homogeneous cell population. This variability can be
due to deterministic or stochastic events that regulate gene expression and has
been reported to increase prior to cell phate decisions \citep{Mojtahedi2016} as well
as throughout ageing \citep{Martinez-jimenez2017}.

This article complements existing workflows that use the Bioconductor package
ecosystem to analyse scRNA-seq datasets \citep{Lun2016, Kim2019}, including the use
of \emph{\href{https://bioconductor.org/packages/3.11/scater}{scater}} and \emph{\href{https://bioconductor.org/packages/3.11/scran}{scran}} to perform quality control
steps and low-level preliminary analysis \citep{McCarthy2017, Lun2016}.
We present a step-by-step computational workflow to robustly quantify
transcriptional variability using the \emph{\href{https://bioconductor.org/packages/3.11/BASiCS}{BASiCS}} package
\citep{Vallejos2015, Vallejos2016, Eling2017}.
\emph{\href{https://bioconductor.org/packages/3.11/BASiCS}{BASiCS}} implements a Bayesian hierarchical framework that
simultaneously performs data normalisation (global scaling), technical noise quantification and selected downstream analyses whilst propagating
statistical uncertainty across these steps.
Within a population of cells, \emph{\href{https://bioconductor.org/packages/3.11/BASiCS}{BASiCS}} decomposes the total
observed variability in gene expression measurements into technical and
biological components. This enables the identification of highly variable genes
that {[}TBC{]}. Moreover, this variance decomposition enables detection of lowly
variable genes with stable expression {[}TBC - CITE GIGASCIENCE PAPER{]}. When two
or more groups of cells are available (e.g.~experimental conditions or
cell types), \emph{\href{https://bioconductor.org/packages/3.11/BASiCS}{BASiCS}} uses differential expression analysis to identify genes whose expression patterns change \citep{Vallejos2016}.

Since the era of RNA sequencing, methods for differential expression testing
of transcript counts across conditions have been developed
\citep{Anders2010, Robinson2009}.
Due to high technical variability and sparsity in scRNA-seq data, new
approaches were developed for differential expression testing
for scRNA-seq data \citep{Katayama2013, Kharchenko2014, Delmans2016}.
In contrast to bulk samples, scRNA-seq measures variations in gene expression
across a population of cells, and can therefore be used to test for changes in
expression variability between two conditions.
To do this, \texttt{BASiCS} compares the gene-specific over-dispersion parameters
between two conditions. These parameters are independent of technical noise
and can be used as proxy for biological variability \citep{Vallejos2016}.
Similar to the mean-variability trend observed for normalised scRNA-seq data
\citep{Brennecke2013}, the estimates for over-dispersion parameters decrease with
mean expression \citep{Vallejos2016}.
To correct for this, BASiCS has been extended to model the mean-variability
relationship and capture residual over-dispersion estimates that show no
association to mean expression.
Therefore, this extension allows to test changes in mean expression in parallel
to changes in variability \citep{Eling2018}.

Two case studies exemplify the use of \texttt{BASiCS} for non-UMI and UMI
scRNA-seq data. In the first case, \texttt{BASiCS} can be used to detect highly and
lowly variable genes and to obtain robust, gene-specific estimates to assess
biological variability in naive CD4\textsuperscript{+} T cells
\citep{Martinez-jimenez2017}; for a similar workflow see \citep{Kim2019}.
Furthermore, we compare naive to activated CD4\textsuperscript{+} T cells to highlight the use
of \texttt{BASiCS} to test for changes in mean expression and expression variability.
In the second case, we use droplet-based scRNA-seq data to detect more subtle
transcriptional changes during embryonic somitogenesis \citep{Ibarra-Soria2018}

\newpage

{\small\bibliography{Workflow.bib}}

\end{document}
